{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/sanjaynagi/AgamPrimer/blob/main/Primer-Design-in-Anopheles-gambiae-dev.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "id": "AM9EeHef6tAD",
      "metadata": {
        "id": "AM9EeHef6tAD",
        "outputId": "03eed63e-35a1-45d5-a570-7f50700a3961",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Installing backend dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "    Preparing wheel metadata ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[K     |████████████████████████████████| 397 kB 5.2 MB/s \n",
            "\u001b[K     |████████████████████████████████| 71 kB 343 kB/s \n",
            "\u001b[K     |████████████████████████████████| 2.3 MB 52.3 MB/s \n",
            "\u001b[K     |████████████████████████████████| 185 kB 57.2 MB/s \n",
            "\u001b[K     |████████████████████████████████| 3.3 MB 38.6 MB/s \n",
            "\u001b[K     |████████████████████████████████| 5.7 MB 60.3 MB/s \n",
            "\u001b[K     |████████████████████████████████| 301 kB 76.3 MB/s \n",
            "\u001b[K     |████████████████████████████████| 123 kB 73.5 MB/s \n",
            "\u001b[K     |████████████████████████████████| 1.6 MB 57.8 MB/s \n",
            "\u001b[K     |████████████████████████████████| 245 kB 52.3 MB/s \n",
            "\u001b[K     |████████████████████████████████| 1.6 MB 58.7 MB/s \n",
            "\u001b[K     |████████████████████████████████| 6.6 MB 53.6 MB/s \n",
            "\u001b[?25h  Building wheel for AgamPrimer (PEP 517) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for primer3-py (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for asciitree (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
          ]
        }
      ],
      "source": [
        "# First, install some packages we require\n",
        "!pip install AgamPrimer==0.3.4 primer3-py malariagen_data -q "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "id": "528ae85d",
      "metadata": {
        "id": "528ae85d"
      },
      "outputs": [],
      "source": [
        "# Import libraries \n",
        "import AgamPrimer\n",
        "import pandas as pd\n",
        "import malariagen_data\n",
        "import numpy as np\n",
        "import primer3\n",
        "import matplotlib.pyplot as plt"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "47bde61a",
      "metadata": {
        "id": "47bde61a"
      },
      "source": [
        "##**AgamPrimer: Primer design considering genetic variation in *Anopheles gambiae***\n",
        "**Author**: [Sanjay Curtis Nagi](https://sanjaynagi.github.io/)    \n",
        "**Email**: sanjay.nagi@lstmed.ac.uk  \n",
        "\n",
        "Often, we would like to design primers for PCR applications, such as genotyping (PCR, TaqMan, LNA) or gene expression qPCR (SYBR, TaqMan, LNA). However, single nucleotide polymorphisms (SNPs) in primer binding sites can result in differences or failures in PCR amplification, referred to as null alleles. \n",
        "\n",
        "In general, mismatches caused by SNPs are more of a problem as you move towards the 3' end. I recommend reading a really good article on this topic on the IDT website - [Consider SNPs when designing PCR and qPCR assays](https://eu.idtdna.com/pages/education/decoded/article/considering-snps-when-designing-pcr-and-qpcr-assays). In *An. gambiae s.l*, there is a [huge amount of genetic variation](https://genome.cshlp.org/content/30/10/1533.full); a SNP found approximately every 1.9 bases (!), which makes considering SNPs even more important when designing molecular assays. Thanks to primer3-py and the fantastic malariagen_data package, we can do all of this in the cloud, hosted by google!  \n",
        "\n",
        "\\\n",
        "\n",
        "####**Google Colab**\n",
        "\n",
        "If you are unfamiliar with iPython notebooks and google colab, I encourage you to read the [FAQ](https://research.google.com/colaboratory/faq.html) and watch the following [introduction](https://www.youtube.com/watch?v=inN8seMm7UI). In general, the cells contain python code, and can be run by pressing the play button next to each cell, and should be run in order.\n",
        "\n",
        "You may want to save a copy of the notebook into your google drive (`file -> save a copy in Drive`)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 81,
      "id": "SV03ewwc9de8",
      "metadata": {
        "id": "SV03ewwc9de8",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title **Selecting Primer Parameters** { run: \"auto\" }\n",
        "#@markdown In the below cells, replace the values with those desired for your primers.\n",
        "\n",
        "assay_type = 'qPCR primers'           #@param [\"gDNA primers\", \"gDNA primers + probe\", \"probe\", \"qPCR primers\"]\n",
        "assay_name = 'coeae1f'        #@param {type:\"string\"}\n",
        "min_amplicon_size = 60        #@param {type:\"integer\"}\n",
        "max_amplicon_size = 120      #@param {type:\"integer\"}\n",
        "amplicon_size_range = [[min_amplicon_size, max_amplicon_size]]\n",
        "n_primer_pairs = 6            #@param {type:\"slider\", min:1, max:20, step:1}\n",
        "\n",
        "#@markdown    \n",
        "#@markdown target_loc is required for gDNA primers and probes, and transcript required for qPCR primers.\n",
        "contig =  \"2L\"                #@param ['2L', '2R', '3L', '3R', 'X']\n",
        "\n",
        "target_loc =  '0'                #@param {type:\"string\"}\n",
        "target_loc = int(target_loc)\n",
        "transcript =  'AGAP006227-RA'         #@param {type:\"string\"} \n",
        "\n",
        "\n",
        "if any(item in assay_type for item in ['gDNA', 'probe']):\n",
        "  assert target_loc > 0, \"Target location must be above 0 and less than the contig length\"\n",
        "elif assay_type == 'qPCR primers':\n",
        "  assert len(transcript) > 2, \"Transcript ID is not valid, should be vectorbase ID such as 'AGAP004707-RD'\""
      ]
    },
    {
      "cell_type": "markdown",
      "id": "aa733775",
      "metadata": {
        "id": "aa733775"
      },
      "source": [
        "Load sequence data for chromosomal arm of choice, using the [malariagen_data API](https://malariagen.github.io/vector-data/ag3/api.html):"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 82,
      "id": "5fa0be8d",
      "metadata": {
        "id": "5fa0be8d",
        "outputId": "08dc89a2-7b74-482a-a40c-6b039c56326c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Our genome sequence for 2L is 49364325 bp long\n"
          ]
        }
      ],
      "source": [
        "# Connect to the malariagen_data ag3 API\n",
        "ag3 = malariagen_data.Ag3() #pre=True\n",
        "genome_seq = ag3.genome_sequence(region=contig)\n",
        "print(f\"Our genome sequence for {contig} is {genome_seq.shape[0]} bp long\")"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "144a37ce",
      "metadata": {
        "id": "144a37ce"
      },
      "source": [
        "Now we need to extract the bit of sequence we need. We will use functions in the [AgamPrimer](https://pypi.org/project/AgamPrimer/) package."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 83,
      "id": "R0CkEd38VXGY",
      "metadata": {
        "id": "R0CkEd38VXGY",
        "outputId": "2816b544-9cfa-472f-a041-1bc089bb4d8b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:geneset: handle region\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Exon junctions for AGAP006227-RA: [ 375  629  973 1364 1514 1651] [28545771, 28546096, 28546517, 28546975, 28547196, 28547556] \n",
            "\n"
          ]
        }
      ],
      "source": [
        "if any(item in assay_type for item in ['gDNA', 'probe']):\n",
        "  # genomic DNA\n",
        "  target_sequence, gdna_pos, seq_parameters = AgamPrimer.prepare_gDNA_sequence(target_loc=target_loc, amplicon_size_range=amplicon_size_range, genome_seq=genome_seq, assay_name=assay_name, assay_type=assay_type)\n",
        "elif assay_type == 'qPCR primers':\n",
        "  # quantitative PCR \n",
        "  target_sequence, exon_junctions, gdna_pos, seq_parameters = AgamPrimer.prepare_cDNA_sequence(transcript=transcript, gff=ag3.geneset(), genome_seq=genome_seq, assay_name=assay_name)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "_aaCJGAXUCSx",
      "metadata": {
        "id": "_aaCJGAXUCSx"
      },
      "source": [
        "Now we have our target sequence. Lets take a look..."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 84,
      "id": "nDaWXq7f9hyA",
      "metadata": {
        "id": "nDaWXq7f9hyA",
        "outputId": "e9fc0643-134d-41ad-ad6a-23988eccfd9a",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 113
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'AATTGTTTAAGTTTCTTGTAAATTTATCTTTAAAGCAACCAACAACGACGTACATGAAAAATCTAATAATTTGATTTATCAATTTAATAAAAAATATAACTACGATATTTAGTAAGCTACTAAAGATGCAATCAAACTGTACTACCAGCACCGTATAAAGTGTTTACCTCACTTAATACAACTGTTCCCGTTTGTACAACGAATCCCACAGCAGCATGCGTCTCATTTCCGGCAAATCGATAAAGCTCACACCGCCATTGTTAATATTTAAACACTTGTACGGCTCCACCTTTGCACCGACCGGTTGCCATATATCGCCAACACTGCCGCCAACATTCGGGCTGCCCACGGTCGCAAAGTTAGTAAACAACTCGACCAGCGTTTGCATCGTCCTGTACTCCATCGTGCCCACATCCGGCACCGGGCTGAACACGTTCTTAAACAGATACGAAAGATCATCCGCATGGGCGGTACCGCGCACGTTACGATCGCAAAAATAGATGCGATAGTGATTGTACGTCTCCGAATCCACTGCAAACCGGTAGACGAACGTTTTGGCCGGCTTTTGGGAGCTGATCCTCGAGCAGACGGTACGGTGCAGCCCGTGCCAGAAGAGCTTATCGGTCATGAGCGTTAAATAGCCTTCCCGATTGTCGAACGATGGTTGGGATCCACCGTAGTAGAACTTTTTCAACTTCAGCCCATACTGCTGGCATGCTTGCGATCCACGCACCAGCTCTAGCTCGGTAGGTACGAGATACTCGAAGTTTTTCAGATTATCCATTATGGCGGGGCTTTCCTTGATGGAGCTCAAGCAGAACAATCCTTCCTCGGCGTTTCCCCCGATCAGTATGTCGATGTCGTTGCTCCATGCCTCACGGCACATTTCTAATGGTGCCTTCGGAATGAAAGTGCTTTCTGTTACGTACGGTTCCACCACTGGACCAAAAGCGAACAGAATACGGTTTTCCAGCTCATTCTTAGTTAGTAGCAGATCCTGATGCTTAATAATCTCCTCGGCCGGTGTTGCTCTCAGCACCTTCAGCAATTCCGCTTCGCCTCCCTTCCCGTTCCAACCGAGCGCTTTGGCCAACCTTTCCGCCATACCGCGTCTCGGCGCCGTTGACCAGTTGTTCAACGTACAGCCCGACATCACTATCGCCCGCCGGAACAACTGTCTCGATTGGCTCGATACCATATGGTAATGGACCGAGCAGCCACCGGCACTCTCACCGAACAGCGTCACATTGTCGGGATCGCCACCGAACTGTGCGATGTTGTCGCGCACCCAGCGAAGTGCCACATTCTGATCTTTCAAACCACCATTTCCGGGTATGCCAAGCTCCTTCGAATCGAGGGACAAAAATCCGAAAGCTCCAATGCGATAGTTGAACGTAACCAGCACAATGTCCTTTTGGATGAGATAGTCCGGTCCGTACATTTCGGTTCCGCTTGAGCCGCGCATAAAAGCGCCCCCATGAATCCAAACCATTACCGGGCGCAATTTTGCCGGTTTGAGCTCTTTCGTGTACACATTCATGTACAGGTGATCCTCACTGCCCACGATCTTGTTCAGCAACTTGTTGAACTGATATCCACTGGGGCCCTGCTGGGTGCAGTCCAGCGGCTCCGTCCATGGCTCTGGCAGCTCCAGATCCTTAAAGCGTCGCTCACCGATCGGTGGTTTACAGTAAGGTATACGCTGGAAGTTCAAGTACTCGACACCGGTGGCAGCCGTTTTCTTCACGCCTCGCACTGGGCCGTACTGCGTCACAGCGACAGGGTGATCGGGCTGGCTCATTTTGAATACGTTAGCAAACAACAACCGGAGCAGATCACGTTTGAAACACGTCTGTGAGACCGCACCACACCGCTAGTAAGGTGGCGGAAAGATACCTTTGCTCGGGAAGTTGGTAGCTGGTCCACGGTTATGGCTGAATGTAATCTAGCGG'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 84
        }
      ],
      "source": [
        "target_sequence"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "38fe2530",
      "metadata": {
        "id": "38fe2530"
      },
      "source": [
        "We need to set up some python dictionaries containing our sequence and primer parameters, this will be our input to primer3. In the below cell, you can modify or add primer3 parameters, such as optimal TM, GC content etc etc. A full list of possible parameters and their functions can be found in the [primer3 manual](https://primer3.org/manual.html)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 85,
      "id": "622ac6d9",
      "metadata": {
        "id": "622ac6d9"
      },
      "outputs": [],
      "source": [
        "primer_parameters  =  {\n",
        "        'PRIMER_NUM_RETURN':n_primer_pairs,\n",
        "        'PRIMER_OPT_SIZE': 20,\n",
        "        'PRIMER_TASK':'generic',\n",
        "        'PRIMER_MIN_SIZE': 17,\n",
        "        'PRIMER_MAX_SIZE': 24,\n",
        "        'PRIMER_OPT_TM': 60.0,\n",
        "        'PRIMER_MIN_TM': 57.0,\n",
        "        'PRIMER_MAX_TM': 63.0,\n",
        "        'PRIMER_MIN_GC': 30.0,\n",
        "        'PRIMER_MAX_GC': 75.0,\n",
        "        'PRIMER_PRODUCT_SIZE_RANGE': amplicon_size_range,\n",
        "        'PRIMER_MIN_THREE_PRIME_DISTANCE':3,          # this parameter is the minimum distance between successive pairs. If 1, it means successive primer pairs could be identical bar one base shift\n",
        "        'PRIMER_INTERNAL_OPT_SIZE': 16,               # Probe size preferences if selected, otherwise ignored\n",
        "        'PRIMER_INTERNAL_MIN_SIZE': 10,\n",
        "        'PRIMER_INTERNAL_MAX_SIZE': 22,\n",
        "        'PRIMER_INTERNAL_MIN_TM': 45,\n",
        "        'PRIMER_INTERNAL_MAX_TM':65,                # Probe considerations are quite relaxed, assumed that LNAs will be used later to affect TM\n",
        "        # Extra primer3 parameters can go here\n",
        "        # In the same format as above                       \n",
        "    }\n",
        "\n",
        "primer_parameters = AgamPrimer.primer_params(primer_parameters, assay_type) ## adds some parameters depending on assay type"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0819ba76",
      "metadata": {
        "id": "0819ba76"
      },
      "source": [
        "#### **Run the primer3 algorithm!**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 86,
      "id": "ae6e7525",
      "metadata": {
        "id": "ae6e7525",
        "scrolled": false
      },
      "outputs": [],
      "source": [
        "primer_dict = primer3.designPrimers(seq_args=seq_parameters, global_args=primer_parameters)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "56945ca6",
      "metadata": {
        "id": "56945ca6"
      },
      "source": [
        "It should be *fast*. The output, which we call 'primer_dict', is a python dictionary containing the full results from primer3. We will turn this into a pandas dataframe (i.e a useful, pretty table), containing just the necessary bits of information. First, we'll print some information from the primer3 run."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 87,
      "id": "1bb40846",
      "metadata": {
        "id": "1bb40846",
        "outputId": "6b9c322c-fd50-41d2-9281-46b3d32ec5e7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PRIMER_LEFT_EXPLAIN  :  considered 15144, GC content failed 1124, low tm 4710, high tm 4177, high hairpin stability 11, ok 5122 \n",
            "\n",
            "PRIMER_RIGHT_EXPLAIN  :  considered 15144, GC content failed 931, low tm 4764, high tm 4238, high hairpin stability 3, ok 5208 \n",
            "\n",
            "PRIMER_PAIR_EXPLAIN  :  considered 216846, unacceptable product size 6304, no overlap of required point 210531, primer in pair overlaps a primer in a better pair 5442, ok 11 \n",
            "\n",
            "PRIMER_LEFT_NUM_RETURNED  :  6 \n",
            "\n",
            "PRIMER_RIGHT_NUM_RETURNED  :  6 \n",
            "\n",
            "PRIMER_INTERNAL_NUM_RETURNED  :  0 \n",
            "\n",
            "PRIMER_PAIR_NUM_RETURNED  :  6 \n",
            "\n"
          ]
        }
      ],
      "source": [
        "AgamPrimer.primer3_run_statistics(primer_dict, assay_type)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "CbVMwOdCrNMf",
      "metadata": {
        "id": "CbVMwOdCrNMf"
      },
      "source": [
        "Now lets wrangle this into an easy to look at table."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 88,
      "id": "bf8c5970",
      "metadata": {
        "id": "bf8c5970",
        "outputId": "c05fd498-8550-4d3a-cf11-2fa64542c34d",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 344
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "7\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "primer_pair                                  0                     1  \\\n",
              "parameter                                                              \n",
              "PRIMER_LEFT_SEQUENCE      AGTCCGGTCCGTACATTTCG  CACTTGTACGGCTCCACCTT   \n",
              "PRIMER_RIGHT_SEQUENCE     CACGAAAGAGCTCAAACCGG  CAAACGCTGGTCGAGTTGTT   \n",
              "PRIMER_LEFT_TM                       59.828097             59.965104   \n",
              "PRIMER_RIGHT_TM                      59.487048             59.344267   \n",
              "PRIMER_LEFT_GC_PERCENT                    55.0                  55.0   \n",
              "PRIMER_RIGHT_GC_PERCENT                   55.0                  50.0   \n",
              "PRIMER_LEFT                         (1425, 20)             (272, 20)   \n",
              "PRIMER_RIGHT                        (1528, 20)             (384, 20)   \n",
              "PRIMER_PAIR_PRODUCT_SIZE                   104                   113   \n",
              "\n",
              "primer_pair                                  2                     3  \\\n",
              "parameter                                                              \n",
              "PRIMER_LEFT_SEQUENCE      AGCTTATCGGTCATGAGCGT  AACAACTCGACCAGCGTTTG   \n",
              "PRIMER_RIGHT_SEQUENCE     GGTGGATCCCAACCATCGTT  TAAGAACGTGTTCAGCCCGG   \n",
              "PRIMER_LEFT_TM                       59.252745             59.344267   \n",
              "PRIMER_RIGHT_TM                      60.034477             60.320073   \n",
              "PRIMER_LEFT_GC_PERCENT                    50.0                  50.0   \n",
              "PRIMER_RIGHT_GC_PERCENT                   55.0                  55.0   \n",
              "PRIMER_LEFT                          (614, 20)             (365, 20)   \n",
              "PRIMER_RIGHT                         (675, 20)             (439, 20)   \n",
              "PRIMER_PAIR_PRODUCT_SIZE                    62                    75   \n",
              "\n",
              "primer_pair                                  4                     5  \n",
              "parameter                                                             \n",
              "PRIMER_LEFT_SEQUENCE      TTGCCGGTTTGAGCTCTTTC  CATTTCGGTTCCGCTTGAGC  \n",
              "PRIMER_RIGHT_SEQUENCE     ACAAGATCGTGGGCAGTGAG  GAAAGAGCTCAAACCGGCAA  \n",
              "PRIMER_LEFT_TM                       59.048442             60.179189  \n",
              "PRIMER_RIGHT_TM                      60.036952             59.048442  \n",
              "PRIMER_LEFT_GC_PERCENT                    50.0                  55.0  \n",
              "PRIMER_RIGHT_GC_PERCENT                   55.0                  50.0  \n",
              "PRIMER_LEFT                         (1506, 20)            (1438, 20)  \n",
              "PRIMER_RIGHT                        (1571, 20)            (1525, 20)  \n",
              "PRIMER_PAIR_PRODUCT_SIZE                    66                    88  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e82ec04d-46f0-46bf-b34f-4f324b022a22\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>primer_pair</th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "      <th>2</th>\n",
              "      <th>3</th>\n",
              "      <th>4</th>\n",
              "      <th>5</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>parameter</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>PRIMER_LEFT_SEQUENCE</th>\n",
              "      <td>AGTCCGGTCCGTACATTTCG</td>\n",
              "      <td>CACTTGTACGGCTCCACCTT</td>\n",
              "      <td>AGCTTATCGGTCATGAGCGT</td>\n",
              "      <td>AACAACTCGACCAGCGTTTG</td>\n",
              "      <td>TTGCCGGTTTGAGCTCTTTC</td>\n",
              "      <td>CATTTCGGTTCCGCTTGAGC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_RIGHT_SEQUENCE</th>\n",
              "      <td>CACGAAAGAGCTCAAACCGG</td>\n",
              "      <td>CAAACGCTGGTCGAGTTGTT</td>\n",
              "      <td>GGTGGATCCCAACCATCGTT</td>\n",
              "      <td>TAAGAACGTGTTCAGCCCGG</td>\n",
              "      <td>ACAAGATCGTGGGCAGTGAG</td>\n",
              "      <td>GAAAGAGCTCAAACCGGCAA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_LEFT_TM</th>\n",
              "      <td>59.828097</td>\n",
              "      <td>59.965104</td>\n",
              "      <td>59.252745</td>\n",
              "      <td>59.344267</td>\n",
              "      <td>59.048442</td>\n",
              "      <td>60.179189</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_RIGHT_TM</th>\n",
              "      <td>59.487048</td>\n",
              "      <td>59.344267</td>\n",
              "      <td>60.034477</td>\n",
              "      <td>60.320073</td>\n",
              "      <td>60.036952</td>\n",
              "      <td>59.048442</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_LEFT_GC_PERCENT</th>\n",
              "      <td>55.0</td>\n",
              "      <td>55.0</td>\n",
              "      <td>50.0</td>\n",
              "      <td>50.0</td>\n",
              "      <td>50.0</td>\n",
              "      <td>55.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_RIGHT_GC_PERCENT</th>\n",
              "      <td>55.0</td>\n",
              "      <td>50.0</td>\n",
              "      <td>55.0</td>\n",
              "      <td>55.0</td>\n",
              "      <td>55.0</td>\n",
              "      <td>50.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_LEFT</th>\n",
              "      <td>(1425, 20)</td>\n",
              "      <td>(272, 20)</td>\n",
              "      <td>(614, 20)</td>\n",
              "      <td>(365, 20)</td>\n",
              "      <td>(1506, 20)</td>\n",
              "      <td>(1438, 20)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_RIGHT</th>\n",
              "      <td>(1528, 20)</td>\n",
              "      <td>(384, 20)</td>\n",
              "      <td>(675, 20)</td>\n",
              "      <td>(439, 20)</td>\n",
              "      <td>(1571, 20)</td>\n",
              "      <td>(1525, 20)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PRIMER_PAIR_PRODUCT_SIZE</th>\n",
              "      <td>104</td>\n",
              "      <td>113</td>\n",
              "      <td>62</td>\n",
              "      <td>75</td>\n",
              "      <td>66</td>\n",
              "      <td>88</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e82ec04d-46f0-46bf-b34f-4f324b022a22')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-e82ec04d-46f0-46bf-b34f-4f324b022a22 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-e82ec04d-46f0-46bf-b34f-4f324b022a22');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 88
        }
      ],
      "source": [
        "primer_df = AgamPrimer.primer3_to_pandas(primer_dict, assay_type)\n",
        "primer_df"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "Lf0CeEVrtaxP",
      "metadata": {
        "id": "Lf0CeEVrtaxP"
      },
      "source": [
        "\n",
        "We can write this to .tsv and excel files, which can be explored in other editors. To download a file from colab to your local computer, click the folder panel on the left-hand sidebar, the three dots next your primers.tsv/.xlsx file, and download."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 89,
      "id": "09eQYSeT7lHG",
      "metadata": {
        "id": "09eQYSeT7lHG"
      },
      "outputs": [],
      "source": [
        "primer_df.to_csv(f\"{assay_name}.{assay_type}.primers.tsv\", sep=\"\\t\")\n",
        "primer_df.to_excel(f\"{assay_name}.{assay_type}.primers.xlsx\")"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "b4a9776c",
      "metadata": {
        "id": "b4a9776c"
      },
      "source": [
        "##**Looking for variation using the ag1000g resource and malariagen API**\n",
        "\n",
        "In Ag3, samples are organised into sample sets. We can load any sample set from the Ag3 resource, but there are quite a few! Lets look at what each sample set contains, breaking it down by species, year and country. "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 90,
      "id": "61286685",
      "metadata": {
        "id": "61286685",
        "outputId": "ed4fb45d-0f15-424d-834e-26da3967748e",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 49,
          "referenced_widgets": [
            "1c032b889e4348b997161fd4a9f4e113",
            "e6904bebdccc4e4689867a6826a46fd2",
            "ef41c110036f497fa80560c554b01cd5",
            "ede22965068044cca74f1b09c31b8ed9",
            "60e494897ca448cbb59e98dd440f76b7",
            "3ce652ccf9984bb7817df8b110097b34",
            "2d36054a85bb4d92b97d38f71f38207d",
            "c4e3923c21a9460d83ff7914623db3a6",
            "14df1c19ee744bf38b7e131c3f5e79ae",
            "2d294c504fdc4d9480fe90b3465b4fdf",
            "ecc9e37d989b4651962ff368a1659340"
          ]
        }
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Load sample metadata:   0%|          | 0/28 [00:00<?, ?it/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "1c032b889e4348b997161fd4a9f4e113"
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "metadata = ag3.sample_metadata()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 91,
      "id": "iSTRbVpxD3oy",
      "metadata": {
        "id": "iSTRbVpxD3oy",
        "outputId": "83b010e5-d7fc-4c97-f6e4-f50dafb12a74",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "taxon                                                arabiensis  coluzzii  \\\n",
              "sample_set   year  country                                                  \n",
              "AG1000G-AO    2009 Angola                                     0        81   \n",
              "AG1000G-BF-A  2012 Burkina Faso                               0        82   \n",
              "AG1000G-BF-B  2014 Burkina Faso                               3        53   \n",
              "AG1000G-BF-C  2004 Burkina Faso                               0         0   \n",
              "AG1000G-CD    2015 Democratic Republic of the Congo           0         0   \n",
              "AG1000G-CF    1993 Central African Republic                   0         5   \n",
              "              1994 Central African Republic                   0        13   \n",
              "AG1000G-CI    2012 Cote d'Ivoire                              0        80   \n",
              "AG1000G-CM-A  2009 Cameroon                                   0         0   \n",
              "AG1000G-CM-B  2005 Cameroon                                   0         7   \n",
              "AG1000G-CM-C  2013 Cameroon                                   2        19   \n",
              "AG1000G-FR    2011 Mayotte                                    0         0   \n",
              "AG1000G-GA-A  2000 Gabon                                      0         0   \n",
              "AG1000G-GH    2012 Ghana                                      0        64   \n",
              "AG1000G-GM-A  2011 Gambia, The                                0         0   \n",
              "AG1000G-GM-B  2012 Gambia, The                                0         2   \n",
              "AG1000G-GM-C  2012 Gambia, The                                0        12   \n",
              "AG1000G-GN-A  2012 Guinea                                     0         4   \n",
              "AG1000G-GN-B  2012 Guinea                                     0         7   \n",
              "                   Mali                                       0        28   \n",
              "AG1000G-GQ    2002 Equatorial Guinea                          0         0   \n",
              "AG1000G-GW    2010 Guinea-Bissau                              0         0   \n",
              "AG1000G-KE    2000 Kenya                                      0         0   \n",
              "              2007 Kenya                                      3         0   \n",
              "              2012 Kenya                                     10         0   \n",
              "AG1000G-ML-A  2014 Mali                                       0        27   \n",
              "AG1000G-ML-B  2004 Mali                                       2        36   \n",
              "AG1000G-MW    2015 Malawi                                    41         0   \n",
              "AG1000G-MZ    2003 Mozambique                                 0         0   \n",
              "              2004 Mozambique                                 0         0   \n",
              "AG1000G-TZ    2012 Tanzania                                  87         0   \n",
              "              2013 Tanzania                                   1         0   \n",
              "              2015 Tanzania                                 137         0   \n",
              "AG1000G-UG    2012 Uganda                                    82         0   \n",
              "AG1000G-X    -1    Lab Cross                                  0        74   \n",
              "\n",
              "taxon                                                gambiae  gcx1  gcx2  \\\n",
              "sample_set   year  country                                                 \n",
              "AG1000G-AO    2009 Angola                                  0     0     0   \n",
              "AG1000G-BF-A  2012 Burkina Faso                           98     0     0   \n",
              "AG1000G-BF-B  2014 Burkina Faso                           46     0     0   \n",
              "AG1000G-BF-C  2004 Burkina Faso                           13     0     0   \n",
              "AG1000G-CD    2015 Democratic Republic of the Congo       76     0     0   \n",
              "AG1000G-CF    1993 Central African Republic                2     0     0   \n",
              "              1994 Central African Republic               53     0     0   \n",
              "AG1000G-CI    2012 Cote d'Ivoire                           0     0     0   \n",
              "AG1000G-CM-A  2009 Cameroon                              303     0     0   \n",
              "AG1000G-CM-B  2005 Cameroon                               90     0     0   \n",
              "AG1000G-CM-C  2013 Cameroon                               23     0     0   \n",
              "AG1000G-FR    2011 Mayotte                                23     0     0   \n",
              "AG1000G-GA-A  2000 Gabon                                  69     0     0   \n",
              "AG1000G-GH    2012 Ghana                                  36     0     0   \n",
              "AG1000G-GM-A  2011 Gambia, The                             0    68     6   \n",
              "AG1000G-GM-B  2012 Gambia, The                             0     9    19   \n",
              "AG1000G-GM-C  2012 Gambia, The                             2     0   160   \n",
              "AG1000G-GN-A  2012 Guinea                                 40     0     0   \n",
              "AG1000G-GN-B  2012 Guinea                                 83     0     0   \n",
              "                   Mali                                   65     0     0   \n",
              "AG1000G-GQ    2002 Equatorial Guinea                      10     0     0   \n",
              "AG1000G-GW    2010 Guinea-Bissau                           8    93     0   \n",
              "AG1000G-KE    2000 Kenya                                  19     0     0   \n",
              "              2007 Kenya                                   0     0     0   \n",
              "              2012 Kenya                                   0     0     0   \n",
              "AG1000G-ML-A  2014 Mali                                   33     0     0   \n",
              "AG1000G-ML-B  2004 Mali                                   33     0     0   \n",
              "AG1000G-MW    2015 Malawi                                  0     0     0   \n",
              "AG1000G-MZ    2003 Mozambique                              3     0     0   \n",
              "              2004 Mozambique                             71     0     0   \n",
              "AG1000G-TZ    2012 Tanzania                                0     0     0   \n",
              "              2013 Tanzania                               32     0     0   \n",
              "              2015 Tanzania                               32     0     0   \n",
              "AG1000G-UG    2012 Uganda                                207     0     0   \n",
              "AG1000G-X    -1    Lab Cross                              49     0     0   \n",
              "\n",
              "taxon                                                gcx3  \\\n",
              "sample_set   year  country                                  \n",
              "AG1000G-AO    2009 Angola                               0   \n",
              "AG1000G-BF-A  2012 Burkina Faso                         0   \n",
              "AG1000G-BF-B  2014 Burkina Faso                         0   \n",
              "AG1000G-BF-C  2004 Burkina Faso                         0   \n",
              "AG1000G-CD    2015 Democratic Republic of the Congo     0   \n",
              "AG1000G-CF    1993 Central African Republic             0   \n",
              "              1994 Central African Republic             0   \n",
              "AG1000G-CI    2012 Cote d'Ivoire                        0   \n",
              "AG1000G-CM-A  2009 Cameroon                             0   \n",
              "AG1000G-CM-B  2005 Cameroon                             0   \n",
              "AG1000G-CM-C  2013 Cameroon                             0   \n",
              "AG1000G-FR    2011 Mayotte                              0   \n",
              "AG1000G-GA-A  2000 Gabon                                0   \n",
              "AG1000G-GH    2012 Ghana                                0   \n",
              "AG1000G-GM-A  2011 Gambia, The                          0   \n",
              "AG1000G-GM-B  2012 Gambia, The                          0   \n",
              "AG1000G-GM-C  2012 Gambia, The                          0   \n",
              "AG1000G-GN-A  2012 Guinea                               0   \n",
              "AG1000G-GN-B  2012 Guinea                               0   \n",
              "                   Mali                                 0   \n",
              "AG1000G-GQ    2002 Equatorial Guinea                    0   \n",
              "AG1000G-GW    2010 Guinea-Bissau                        0   \n",
              "AG1000G-KE    2000 Kenya                                0   \n",
              "              2007 Kenya                                0   \n",
              "              2012 Kenya                               54   \n",
              "AG1000G-ML-A  2014 Mali                                 0   \n",
              "AG1000G-ML-B  2004 Mali                                 0   \n",
              "AG1000G-MW    2015 Malawi                               0   \n",
              "AG1000G-MZ    2003 Mozambique                           0   \n",
              "              2004 Mozambique                           0   \n",
              "AG1000G-TZ    2012 Tanzania                             0   \n",
              "              2013 Tanzania                            10   \n",
              "              2015 Tanzania                             1   \n",
              "AG1000G-UG    2012 Uganda                               0   \n",
              "AG1000G-X    -1    Lab Cross                            0   \n",
              "\n",
              "taxon                                                intermediate_gambcolu_arabiensis  \\\n",
              "sample_set   year  country                                                              \n",
              "AG1000G-AO    2009 Angola                                                           0   \n",
              "AG1000G-BF-A  2012 Burkina Faso                                                     0   \n",
              "AG1000G-BF-B  2014 Burkina Faso                                                     0   \n",
              "AG1000G-BF-C  2004 Burkina Faso                                                     0   \n",
              "AG1000G-CD    2015 Democratic Republic of the Congo                                 0   \n",
              "AG1000G-CF    1993 Central African Republic                                         0   \n",
              "              1994 Central African Republic                                         0   \n",
              "AG1000G-CI    2012 Cote d'Ivoire                                                    0   \n",
              "AG1000G-CM-A  2009 Cameroon                                                         0   \n",
              "AG1000G-CM-B  2005 Cameroon                                                         0   \n",
              "AG1000G-CM-C  2013 Cameroon                                                         0   \n",
              "AG1000G-FR    2011 Mayotte                                                          0   \n",
              "AG1000G-GA-A  2000 Gabon                                                            0   \n",
              "AG1000G-GH    2012 Ghana                                                            0   \n",
              "AG1000G-GM-A  2011 Gambia, The                                                      0   \n",
              "AG1000G-GM-B  2012 Gambia, The                                                      0   \n",
              "AG1000G-GM-C  2012 Gambia, The                                                      0   \n",
              "AG1000G-GN-A  2012 Guinea                                                           0   \n",
              "AG1000G-GN-B  2012 Guinea                                                           0   \n",
              "                   Mali                                                             0   \n",
              "AG1000G-GQ    2002 Equatorial Guinea                                                0   \n",
              "AG1000G-GW    2010 Guinea-Bissau                                                    0   \n",
              "AG1000G-KE    2000 Kenya                                                            0   \n",
              "              2007 Kenya                                                            0   \n",
              "              2012 Kenya                                                            0   \n",
              "AG1000G-ML-A  2014 Mali                                                             0   \n",
              "AG1000G-ML-B  2004 Mali                                                             0   \n",
              "AG1000G-MW    2015 Malawi                                                           0   \n",
              "AG1000G-MZ    2003 Mozambique                                                       0   \n",
              "              2004 Mozambique                                                       0   \n",
              "AG1000G-TZ    2012 Tanzania                                                         0   \n",
              "              2013 Tanzania                                                         0   \n",
              "              2015 Tanzania                                                         0   \n",
              "AG1000G-UG    2012 Uganda                                                           1   \n",
              "AG1000G-X    -1    Lab Cross                                                        0   \n",
              "\n",
              "taxon                                                intermediate_gambiae_coluzzii  \n",
              "sample_set   year  country                                                          \n",
              "AG1000G-AO    2009 Angola                                                        0  \n",
              "AG1000G-BF-A  2012 Burkina Faso                                                  1  \n",
              "AG1000G-BF-B  2014 Burkina Faso                                                  0  \n",
              "AG1000G-BF-C  2004 Burkina Faso                                                  0  \n",
              "AG1000G-CD    2015 Democratic Republic of the Congo                              0  \n",
              "AG1000G-CF    1993 Central African Republic                                      0  \n",
              "              1994 Central African Republic                                      0  \n",
              "AG1000G-CI    2012 Cote d'Ivoire                                                 0  \n",
              "AG1000G-CM-A  2009 Cameroon                                                      0  \n",
              "AG1000G-CM-B  2005 Cameroon                                                      0  \n",
              "AG1000G-CM-C  2013 Cameroon                                                      0  \n",
              "AG1000G-FR    2011 Mayotte                                                       0  \n",
              "AG1000G-GA-A  2000 Gabon                                                         0  \n",
              "AG1000G-GH    2012 Ghana                                                         0  \n",
              "AG1000G-GM-A  2011 Gambia, The                                                   0  \n",
              "AG1000G-GM-B  2012 Gambia, The                                                   1  \n",
              "AG1000G-GM-C  2012 Gambia, The                                                   0  \n",
              "AG1000G-GN-A  2012 Guinea                                                        1  \n",
              "AG1000G-GN-B  2012 Guinea                                                        1  \n",
              "                   Mali                                                          1  \n",
              "AG1000G-GQ    2002 Equatorial Guinea                                             0  \n",
              "AG1000G-GW    2010 Guinea-Bissau                                                 0  \n",
              "AG1000G-KE    2000 Kenya                                                         0  \n",
              "              2007 Kenya                                                         0  \n",
              "              2012 Kenya                                                         0  \n",
              "AG1000G-ML-A  2014 Mali                                                          0  \n",
              "AG1000G-ML-B  2004 Mali                                                          0  \n",
              "AG1000G-MW    2015 Malawi                                                        0  \n",
              "AG1000G-MZ    2003 Mozambique                                                    0  \n",
              "              2004 Mozambique                                                    0  \n",
              "AG1000G-TZ    2012 Tanzania                                                      0  \n",
              "              2013 Tanzania                                                      0  \n",
              "              2015 Tanzania                                                      0  \n",
              "AG1000G-UG    2012 Uganda                                                        0  \n",
              "AG1000G-X    -1    Lab Cross                                                   174  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ca3a1ca7-dd19-4493-a447-da84ff70bfc2\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>taxon</th>\n",
              "      <th>arabiensis</th>\n",
              "      <th>coluzzii</th>\n",
              "      <th>gambiae</th>\n",
              "      <th>gcx1</th>\n",
              "      <th>gcx2</th>\n",
              "      <th>gcx3</th>\n",
              "      <th>intermediate_gambcolu_arabiensis</th>\n",
              "      <th>intermediate_gambiae_coluzzii</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sample_set</th>\n",
              "      <th>year</th>\n",
              "      <th>country</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>AG1000G-AO</th>\n",
              "      <th>2009</th>\n",
              "      <th>Angola</th>\n",
              "      <td>0</td>\n",
              "      <td>81</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-BF-A</th>\n",
              "      <th>2012</th>\n",
              "      <th>Burkina Faso</th>\n",
              "      <td>0</td>\n",
              "      <td>82</td>\n",
              "      <td>98</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-BF-B</th>\n",
              "      <th>2014</th>\n",
              "      <th>Burkina Faso</th>\n",
              "      <td>3</td>\n",
              "      <td>53</td>\n",
              "      <td>46</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-BF-C</th>\n",
              "      <th>2004</th>\n",
              "      <th>Burkina Faso</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>13</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-CD</th>\n",
              "      <th>2015</th>\n",
              "      <th>Democratic Republic of the Congo</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>76</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">AG1000G-CF</th>\n",
              "      <th>1993</th>\n",
              "      <th>Central African Republic</th>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1994</th>\n",
              "      <th>Central African Republic</th>\n",
              "      <td>0</td>\n",
              "      <td>13</td>\n",
              "      <td>53</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-CI</th>\n",
              "      <th>2012</th>\n",
              "      <th>Cote d'Ivoire</th>\n",
              "      <td>0</td>\n",
              "      <td>80</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-CM-A</th>\n",
              "      <th>2009</th>\n",
              "      <th>Cameroon</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>303</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-CM-B</th>\n",
              "      <th>2005</th>\n",
              "      <th>Cameroon</th>\n",
              "      <td>0</td>\n",
              "      <td>7</td>\n",
              "      <td>90</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-CM-C</th>\n",
              "      <th>2013</th>\n",
              "      <th>Cameroon</th>\n",
              "      <td>2</td>\n",
              "      <td>19</td>\n",
              "      <td>23</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-FR</th>\n",
              "      <th>2011</th>\n",
              "      <th>Mayotte</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>23</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GA-A</th>\n",
              "      <th>2000</th>\n",
              "      <th>Gabon</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>69</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GH</th>\n",
              "      <th>2012</th>\n",
              "      <th>Ghana</th>\n",
              "      <td>0</td>\n",
              "      <td>64</td>\n",
              "      <td>36</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GM-A</th>\n",
              "      <th>2011</th>\n",
              "      <th>Gambia, The</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>68</td>\n",
              "      <td>6</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GM-B</th>\n",
              "      <th>2012</th>\n",
              "      <th>Gambia, The</th>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>9</td>\n",
              "      <td>19</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GM-C</th>\n",
              "      <th>2012</th>\n",
              "      <th>Gambia, The</th>\n",
              "      <td>0</td>\n",
              "      <td>12</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>160</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GN-A</th>\n",
              "      <th>2012</th>\n",
              "      <th>Guinea</th>\n",
              "      <td>0</td>\n",
              "      <td>4</td>\n",
              "      <td>40</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">AG1000G-GN-B</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">2012</th>\n",
              "      <th>Guinea</th>\n",
              "      <td>0</td>\n",
              "      <td>7</td>\n",
              "      <td>83</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Mali</th>\n",
              "      <td>0</td>\n",
              "      <td>28</td>\n",
              "      <td>65</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GQ</th>\n",
              "      <th>2002</th>\n",
              "      <th>Equatorial Guinea</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-GW</th>\n",
              "      <th>2010</th>\n",
              "      <th>Guinea-Bissau</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>8</td>\n",
              "      <td>93</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">AG1000G-KE</th>\n",
              "      <th>2000</th>\n",
              "      <th>Kenya</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>19</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2007</th>\n",
              "      <th>Kenya</th>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2012</th>\n",
              "      <th>Kenya</th>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>54</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-ML-A</th>\n",
              "      <th>2014</th>\n",
              "      <th>Mali</th>\n",
              "      <td>0</td>\n",
              "      <td>27</td>\n",
              "      <td>33</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-ML-B</th>\n",
              "      <th>2004</th>\n",
              "      <th>Mali</th>\n",
              "      <td>2</td>\n",
              "      <td>36</td>\n",
              "      <td>33</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-MW</th>\n",
              "      <th>2015</th>\n",
              "      <th>Malawi</th>\n",
              "      <td>41</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">AG1000G-MZ</th>\n",
              "      <th>2003</th>\n",
              "      <th>Mozambique</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2004</th>\n",
              "      <th>Mozambique</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>71</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">AG1000G-TZ</th>\n",
              "      <th>2012</th>\n",
              "      <th>Tanzania</th>\n",
              "      <td>87</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2013</th>\n",
              "      <th>Tanzania</th>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>32</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2015</th>\n",
              "      <th>Tanzania</th>\n",
              "      <td>137</td>\n",
              "      <td>0</td>\n",
              "      <td>32</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-UG</th>\n",
              "      <th>2012</th>\n",
              "      <th>Uganda</th>\n",
              "      <td>82</td>\n",
              "      <td>0</td>\n",
              "      <td>207</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AG1000G-X</th>\n",
              "      <th>-1</th>\n",
              "      <th>Lab Cross</th>\n",
              "      <td>0</td>\n",
              "      <td>74</td>\n",
              "      <td>49</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>174</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ca3a1ca7-dd19-4493-a447-da84ff70bfc2')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-ca3a1ca7-dd19-4493-a447-da84ff70bfc2 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-ca3a1ca7-dd19-4493-a447-da84ff70bfc2');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 91
        }
      ],
      "source": [
        "pivot_country_year_taxon = (\n",
        "    metadata\n",
        "    .pivot_table(\n",
        "        index=[\"sample_set\", \"year\", \"country\"], \n",
        "        columns=[\"taxon\"], \n",
        "        values=\"sample_id\",\n",
        "        aggfunc=\"count\",\n",
        "        fill_value=0\n",
        "    )\n",
        ")\n",
        "\n",
        "pivot_country_year_taxon"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "BLwvXBT28NYE",
      "metadata": {
        "id": "BLwvXBT28NYE"
      },
      "source": [
        "Here, we can see the breakdown by sample set for country, species and year. For the purposes of this notebook, let's use the Ghana sample set. If we wanted to use all sample sets, we could supply '3.0' instead of a sample set, which will load all samples from the ag3.0 release."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 92,
      "id": "49ca7f2b",
      "metadata": {
        "id": "49ca7f2b"
      },
      "outputs": [],
      "source": [
        "sample_set = 'AG1000G-GH'          # sample_set = '3.0' .you can also supply lists with multiple sample sets e.g ['AG1000G-GH', 'AG1000G-CI', 'AG1000G-BF-A]\n",
        "sample_query = None # \"aim_species == 'gambiae'\"                # here we can subset to specific values in the metadata e.g :     \"aim_species == 'gambiae'\" , or \"aim_species == 'arabiensis'\"          "
      ]
    },
    {
      "cell_type": "markdown",
      "id": "67coxQKoo82x",
      "metadata": {
        "id": "67coxQKoo82x"
      },
      "source": [
        "### **Plot allele frequencies in primers locations**\n",
        "\n",
        "Now we can plot the primers pairs, and the frequency of any alternate alleles in the Ag1000g sample set of choice. When calculating allele frequencies, we will take the sum of all alternate alleles, as we are interested here in any mutations which are different from the reference genome. \n",
        "\n",
        "We will also plot the primer Tm, GC and genomic spans of each primer binding site. We can use this plot to identify primers pairs and probes which may be suitable, particularly trying to avoid SNPs in the 3' end. "
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import allel"
      ],
      "metadata": {
        "id": "z8rHhqpeo1bv"
      },
      "id": "z8rHhqpeo1bv",
      "execution_count": 93,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "span_str=f'{contig}:{gdna_pos.min()}-{gdna_pos.max()}'                        \n",
        "\n",
        "geno = ag3.snp_genotypes(region=span_str, sample_sets=sample_set, sample_query=sample_query) # get genotypes\n",
        "ac = allel.GenotypeDaskArray(geno).count_alleles()"
      ],
      "metadata": {
        "id": "FetjE3XbosQ_",
        "outputId": "69d24f42-4d84-4363-8caa-69c4e6564aa4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "FetjE3XbosQ_",
      "execution_count": 94,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply site filters if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply sample query if requested\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "alts = ag3.snp_sites(region=span_str, field='ALT').compute()"
      ],
      "metadata": {
        "id": "32_Y5t50pA83",
        "outputId": "2bfb8542-9a03-4a9c-a0a2-87e0b33e65bf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "32_Y5t50pA83",
      "execution_count": 95,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "freqs = ac.to_frequencies() \n",
        "freqs = freqs.compute()"
      ],
      "metadata": {
        "id": "1vXvpJk7r6Di"
      },
      "id": "1vXvpJk7r6Di",
      "execution_count": 96,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# if there are no ALTs lets add the zeros for the ALTs otherwise only REF count returned \n",
        "\n",
        "def addZeroCols(freqs):\n",
        "  freqlength = freqs.shape[1]\n",
        "  needed = 4-freqlength\n",
        "  if needed > 0:\n",
        "    for i in range(needed):\n",
        "      freqs = np.column_stack([freqs, np.repeat(0, freqs.shape[0])])\n",
        "  return(freqs)"
      ],
      "metadata": {
        "id": "prCelkDypkfY"
      },
      "id": "prCelkDypkfY",
      "execution_count": 97,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "alts = ag3.snp_sites(region=span_str, field='ALT')\n",
        "refs = ag3.snp_sites(region=span_str, field='REF')\n",
        "ref_alt_array = np.column_stack([refs, alts]).astype('U13')"
      ],
      "metadata": {
        "id": "_1FtzqxSpVlT",
        "outputId": "ba2c3467-e5cb-42fd-9d35-4f4e5e4f7644",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "_1FtzqxSpVlT",
      "execution_count": 99,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import plotly\n",
        "import plotly.express as px"
      ],
      "metadata": {
        "id": "GY1gaXGV7zRE"
      },
      "id": "GY1gaXGV7zRE",
      "execution_count": 100,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "freqs.shape"
      ],
      "metadata": {
        "id": "Xn1S56fm22US",
        "outputId": "80463ee9-26d4-4f3b-8990-fc0b245a26e4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "Xn1S56fm22US",
      "execution_count": 101,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(2543, 4)"
            ]
          },
          "metadata": {},
          "execution_count": 101
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "      snps = ag3.snp_calls(region=span_str, sample_sets=sample_set)     # get genotypes\n"
      ],
      "metadata": {
        "id": "T9MI6SMIw2Vf",
        "outputId": "041d7659-2a38-4935-8107-33985bdf79a7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "T9MI6SMIw2Vf",
      "execution_count": 117,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "def get_base_freqs(freqs, ref_alt_array):\n",
        "  freq_df = pd.DataFrame(ref_alt_array)\n",
        "  for i_base in range(4):\n",
        "    for i in range(freqs.shape[0]):\n",
        "      base = ref_alt_array[i, i_base]\n",
        "      freq_df.loc[i, f'{base}_freq'] = freqs[i, i_base]\n",
        "  return(freq_df)\n",
        "\n",
        "def get_primer_arrays(contig, gdna_pos, sample_set, assay_type, sample_query=None):\n",
        "\n",
        "  if any(item in assay_type for item in ['gDNA', 'probe']):\n",
        "    span_str=f'{contig}:{gdna_pos.min()}-{gdna_pos.max()}'                        \n",
        "    snps = ag3.snp_calls(region=span_str, sample_sets=sample_set)     # get genotypes\n",
        "    ref_alt_arr = snps['variant_allele']\n",
        "    geno = snps['call_genotype']\n",
        "    freq_arr = allel.GenotypeArray(geno).count_alleles().to_frequencies() \n",
        "    pos_arr = gdna_pos\n",
        "  elif assay_type == 'qPCR primers':\n",
        "    freq_arr = []\n",
        "    ref_alt_arr = []\n",
        "    pos_arr = np.array([])\n",
        "    exon_spans = np.array(AgamPrimer.consecutive(gdna_pos)) + 1\n",
        "    for span in exon_spans:\n",
        "      span_str = f'{contig}:{span[0]}-{span[1]}'                        \n",
        "      snps = ag3.snp_calls(region=span_str, sample_sets=sample_set)     # get genotypes\n",
        "      ref_alts = snps['variant_allele']\n",
        "      geno = snps['call_genotype']\n",
        "      freqs = allel.GenotypeArray(geno).count_alleles().to_frequencies()                                # calculate allele frequencies\n",
        "      freqs = addZeroCols(freqs)\n",
        "      freq_arr.append(freqs)\n",
        "      ref_alt_arr.append(ref_alts)\n",
        "      pos_arr = np.append(pos_arr, np.arange(span[0], span[1]+1).astype(int))\n",
        "    freq_arr = np.concatenate(freq_arr)\n",
        "    ref_alt_arr = np.concatenate(ref_alt_arr)\n",
        "\n",
        "  return(freq_arr, ref_alt_arr.astype('U13'), pos_arr)\n",
        "\n",
        "\n",
        "def get_primer_alt_frequencies(primer_df, gdna_pos, pair, sample_set, assay_type, contig, sample_query):\n",
        "  \"\"\"\n",
        "  Find the genomic locations of pairs of primers, and runs span_to_freq\n",
        "  to get allele frequencies at those locations\n",
        "  \"\"\"\n",
        "\n",
        "  oligos, _ = AgamPrimer.return_oligo_list(assay_type)\n",
        "  base_freqs, ref_alt_arr, pos_arr = get_primer_arrays(contig=contig, gdna_pos=gdna_pos, sample_set=sample_set, assay_type=assay_type, sample_query=sample_query)\n",
        "\n",
        "  freq_arr = base_freqs[:, 1:].sum(axis=1)\n",
        "\n",
        "  di = {}\n",
        "  for oligo in oligos:\n",
        "    primer_loc = primer_df.loc[f'PRIMER_{oligo}', str(pair)][0]\n",
        "    primer_loc = primer_loc + 1 if oligo == 'RIGHT' else primer_loc\n",
        "    primer_size = primer_df.loc[f'PRIMER_{oligo}', str(pair)][1]\n",
        "    if oligo in ['LEFT', 'INTERNAL']:\n",
        "      freq = freq_arr[primer_loc:primer_loc+primer_size]\n",
        "      base_freqs_arr = base_freqs[primer_loc:primer_loc+primer_size, :]\n",
        "      ref = ref_alt_arr[primer_loc:primer_loc+primer_size, 0]\n",
        "      ref_alt = ref_alt_arr[primer_loc:primer_loc+primer_size, :]\n",
        "      pos = pos_arr[primer_loc:primer_loc+primer_size]\n",
        "    elif oligo == 'RIGHT':\n",
        "      freq = np.flip(freq_arr[primer_loc-primer_size:primer_loc])\n",
        "      base_freqs_arr = base_freqs[primer_loc-primer_size:primer_loc, :]\n",
        "      base_freqs_arr = np.flip(base_freqs_arr, axis=0)\n",
        "      ref = ref_alt_arr[primer_loc-primer_size:primer_loc, 0] \n",
        "      ref = np.array(list(AgamPrimer.rev_complement(''.join(ref))), dtype=str)\n",
        "      ref_alt = ref_alt_arr[primer_loc-primer_size:primer_loc, :]\n",
        "      pos = np.flip(pos_arr[primer_loc-primer_size:primer_loc])\n",
        "\n",
        "    df = pd.DataFrame({'position': pos, 'base':ref, 'alt_frequency':freq}) # Make dataframe for plotting\n",
        "    df['base_pos'] = df['base'] + \"_\" + df['position'].astype(str)\n",
        "    assert df.shape[0] == primer_size, \"Wrong size primers\"\n",
        "    print(oligo)\n",
        "    freq_df = get_base_freqs(base_freqs_arr, ref_alt).filter(like='freq')\n",
        "    df = pd.concat([df, freq_df], axis=1)\n",
        "    di[oligo] = df\n",
        "  return(di)"
      ],
      "metadata": {
        "id": "yK3KJn980lMJ"
      },
      "id": "yK3KJn980lMJ",
      "execution_count": 245,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def plot_primer_ag3_frequencies(primer_df, gdna_pos, contig, sample_set, assay_type, seq_parameters, save=True, sample_query=None):\n",
        "  \"\"\"\n",
        "  Loop through n primer pairs, retrieving frequency data and plot allele frequencies\n",
        "  \"\"\"\n",
        "\n",
        "  if sample_query != None:\n",
        "    print(f\"Subsetting allele frequencies to {sample_query}\")\n",
        "\n",
        "  n_primer_pairs=len(primer_df.columns)\n",
        "  name = seq_parameters['SEQUENCE_ID']\n",
        "  exon_junctions = seq_parameters['SEQUENCE_OVERLAP_JUNCTION_LIST'] if assay_type == 'qPCR primers' else None\n",
        "  transcript = seq_parameters['TRANSCRIPT'] if assay_type == 'qPCR primers' else None\n",
        "  target_loc = seq_parameters['GENOMIC_SEQUENCE_TARGET'] if any(item in assay_type for item in ['gDNA', 'probe']) else None\n",
        "  res_dict = {}\n",
        "  # Loop through each primer pair and get the frequencies of alternate alleles, storing in dict\n",
        "  for i in range(n_primer_pairs):\n",
        "    res_dict[i] = get_primer_alt_frequencies(primer_df, gdna_pos, i, sample_set, assay_type, contig, sample_query)\n",
        "  \n",
        "  return(primer_df, res_dict, assay_type, exon_junctions, target_loc)\n",
        "\n",
        "  # Plot data\n",
        "  oligos, _ = return_oligo_list(assay_type)\n",
        "  fig, ax = plt.subplots(n_primer_pairs, len(oligos), figsize=[6*len(oligos), (n_primer_pairs*2)+2], constrained_layout=True)    \n",
        "  if any(item in assay_type for item in ['gDNA']):\n",
        "    fig.suptitle(f\"{name} primer pairs | {sample_set} | target {target_loc} bp\", fontweight='bold')\n",
        "  elif assay_type == 'probe':\n",
        "    fig.suptitle(f\"{name} probe | {sample_set} | target {target_loc} bp\", fontweight='bold')\n",
        "  elif assay_type == 'qPCR primers':\n",
        "    fig.suptitle(f\"{name} primer pairs | {sample_set} | target {transcript}\", fontweight='bold')\n",
        "\n",
        "  for i in range(n_primer_pairs):\n",
        "    plot_primer(primer_df, ax, i, res_dict, assay_type, exon_junctions=exon_junctions, target_loc=target_loc)\n",
        "  if save: fig.savefig(f\"{name}.{assay_type}.primers.png\")\n",
        "  return(res_dict)\n",
        "\n",
        "\n",
        "def plot_primer(primer_df, ax, i, res_dict, assay_type, exon_junctions=None, target_loc=None):\n",
        "  \"\"\"\n",
        "  Plot primer allele frequencies and text\n",
        "  \"\"\"\n",
        "  oligos, _ = AgamPrimer.return_oligo_list(assay_type)\n",
        "\n",
        "  for idx, oligo in enumerate(oligos):\n",
        "    if len(oligos) > 1:\n",
        "      axes = ax[i, idx]\n",
        "    else:\n",
        "      axes = ax[i]\n",
        "\n",
        "    sns.scatterplot(ax=axes, x=res_dict[i][oligo]['base_pos'], y=res_dict[i][oligo]['alt_frequency'], s=200) \n",
        "    axes.set_xticklabels(res_dict[i][oligo]['base'])\n",
        "    axes.set_ylim(0,1)\n",
        "    axes.set_xlabel(\"\")\n",
        "    axes.set_ylabel(\"Alternate allele frequency\")\n",
        "    if idx > 0: \n",
        "      axes.set_ylabel(\"\")\n",
        "      axes.set_yticklabels(\"\")\n",
        "\n",
        "    if assay_type == 'qPCR primers':\n",
        "      a = primer_df.loc[f'PRIMER_{oligo}', str(i)]\n",
        "      a = a[0] - a[1] if oligo == 'RIGHT' else a[0] + a[1]\n",
        "      arr = np.array(exon_junctions) - a\n",
        "      if (np.abs(arr) < 18).any():\n",
        "        bases_in_new_exon  = arr[np.abs(arr) < 18]\n",
        "        plt.setp(axes.get_xticklabels()[bases_in_new_exon[0]:], backgroundcolor=\"antiquewhite\") if oligo == 'RIGHT' else plt.setp(axes.get_xticklabels()[:bases_in_new_exon[0]], backgroundcolor=\"antiquewhite\")\n",
        "    \n",
        "    if oligo == 'LEFT':\n",
        "      axes.set_title(f\"Forward primer {i}\") \n",
        "    elif oligo == 'RIGHT':\n",
        "      axes.set_title(f\"Reverse primer {i}\")\n",
        "    elif oligo == 'INTERNAL':\n",
        "      axes.set_title(f\"Probe {i}\")\n",
        "      idx = np.where(res_dict[i][oligo]['position'] == target_loc)[0][0]\n",
        "      plt.setp(axes.get_xticklabels()[idx], backgroundcolor=\"antiquewhite\")\n",
        "    plot_pair_text(primer_df, i, axes, oligo, res_dict)"
      ],
      "metadata": {
        "id": "HeV6p8ZF4g2Q"
      },
      "id": "HeV6p8ZF4g2Q",
      "execution_count": 246,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "primer_df, res_dict, assay_type, exon_junctions, target_loc = plot_primer_ag3_frequencies(primer_df=primer_df,\n",
        "                                                gdna_pos=gdna_pos,\n",
        "                                                contig=contig,\n",
        "                                                sample_set=sample_set, \n",
        "                                                sample_query=sample_query,\n",
        "                                                assay_type=assay_type,\n",
        "                                                seq_parameters=seq_parameters,\n",
        "                                                save=True)"
      ],
      "metadata": {
        "id": "jDKe8dwB9BzK",
        "outputId": "0287488c-dd4a-4897-a9b9-94b2430d9e1c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "jDKe8dwB9BzK",
      "execution_count": 247,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "LEFT\n",
            "RIGHT\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "LEFT\n",
            "RIGHT\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "LEFT\n",
            "RIGHT\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "LEFT\n",
            "RIGHT\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "LEFT\n",
            "RIGHT\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n",
            "DEBUG:malariagen_data.ag3:snp_calls: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: call arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: sample arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_calls_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add variants variables\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_position\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_allele\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: variant_contig\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: site filters arrays\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: set up attributes\n",
            "DEBUG:malariagen_data.ag3:_snp_variants_dataset: create a dataset\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle site class\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle region, do this only once - optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_calls: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_calls: apply site filters\n",
            "DEBUG:malariagen_data.ag3:snp_calls: add call_genotype_mask\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle sample query\n",
            "DEBUG:malariagen_data.ag3:snp_calls: handle cohort size\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "LEFT\n",
            "RIGHT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "res_dict[4]['RIGHT']"
      ],
      "metadata": {
        "id": "x99LksQGkUox",
        "outputId": "d5e3943a-ca37-44f2-8b7b-29a8054b3ac1",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 611
        }
      },
      "id": "x99LksQGkUox",
      "execution_count": 248,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      position base  alt_frequency      base_pos  C_freq  T_freq  A_freq  \\\n",
              "0   28547471.0    A            0.0  A_28547471.0     1.0     0.0     0.0   \n",
              "1   28547470.0    C            0.0  C_28547470.0     0.0     1.0     0.0   \n",
              "2   28547469.0    A            0.0  A_28547469.0     1.0     0.0     0.0   \n",
              "3   28547468.0    A            0.0  A_28547468.0     0.0     0.0     1.0   \n",
              "4   28547467.0    G            0.0  G_28547467.0     1.0     0.0     0.0   \n",
              "5   28547466.0    A            0.0  A_28547466.0     0.0     1.0     0.0   \n",
              "6   28547465.0    T            0.0  T_28547465.0     0.0     0.0     0.0   \n",
              "7   28547464.0    C            0.0  C_28547464.0     1.0     0.0     0.0   \n",
              "8   28547463.0    G            0.0  G_28547463.0     1.0     0.0     0.0   \n",
              "9   28547462.0    T            0.0  T_28547462.0     1.0     0.0     0.0   \n",
              "10  28547461.0    G            0.0  G_28547461.0     0.0     0.0     1.0   \n",
              "11  28547460.0    G            0.0  G_28547460.0     1.0     0.0     0.0   \n",
              "12  28547459.0    G            0.0  G_28547459.0     0.0     0.0     0.0   \n",
              "13  28547458.0    C            0.0  C_28547458.0     0.0     0.0     1.0   \n",
              "14  28547457.0    A            0.0  A_28547457.0     0.0     1.0     0.0   \n",
              "15  28547456.0    G            0.0  G_28547456.0     1.0     0.0     0.0   \n",
              "16  28547455.0    T            0.0  T_28547455.0     0.0     1.0     0.0   \n",
              "17  28547454.0    G            0.0  G_28547454.0     0.0     1.0     0.0   \n",
              "18  28547453.0    A            0.0  A_28547453.0     0.0     0.0     0.0   \n",
              "19  28547452.0    G            0.0  G_28547452.0     0.0     1.0     0.0   \n",
              "\n",
              "    G_freq  \n",
              "0      0.0  \n",
              "1      0.0  \n",
              "2      0.0  \n",
              "3      0.0  \n",
              "4      0.0  \n",
              "5      0.0  \n",
              "6      1.0  \n",
              "7      0.0  \n",
              "8      0.0  \n",
              "9      0.0  \n",
              "10     0.0  \n",
              "11     0.0  \n",
              "12     1.0  \n",
              "13     0.0  \n",
              "14     0.0  \n",
              "15     0.0  \n",
              "16     0.0  \n",
              "17     0.0  \n",
              "18     1.0  \n",
              "19     0.0  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-101ebc25-8b8e-4e2c-a59b-412254eedee9\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>position</th>\n",
              "      <th>base</th>\n",
              "      <th>alt_frequency</th>\n",
              "      <th>base_pos</th>\n",
              "      <th>C_freq</th>\n",
              "      <th>T_freq</th>\n",
              "      <th>A_freq</th>\n",
              "      <th>G_freq</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>28547471.0</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>A_28547471.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>28547470.0</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>C_28547470.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>28547469.0</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>A_28547469.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>28547468.0</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>A_28547468.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>28547467.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547467.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>28547466.0</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>A_28547466.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>28547465.0</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "      <td>T_28547465.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>28547464.0</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>C_28547464.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>28547463.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547463.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>28547462.0</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "      <td>T_28547462.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>28547461.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547461.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>28547460.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547460.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>28547459.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547459.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>28547458.0</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>C_28547458.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>28547457.0</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>A_28547457.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>28547456.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547456.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>28547455.0</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "      <td>T_28547455.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>28547454.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547454.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>28547453.0</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>A_28547453.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>28547452.0</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>G_28547452.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-101ebc25-8b8e-4e2c-a59b-412254eedee9')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-101ebc25-8b8e-4e2c-a59b-412254eedee9 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-101ebc25-8b8e-4e2c-a59b-412254eedee9');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 248
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "lists = res_dict[i][oligo][['A_freq', 'C_freq', 'G_freq', 'T_freq']].to_numpy().T.tolist()"
      ],
      "metadata": {
        "id": "8nchZNu5OILd"
      },
      "id": "8nchZNu5OILd",
      "execution_count": 267,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "if len(oligos) == 2:\n",
        "  plt_title = ['Forward primer', 'Reverse primer']\n",
        "elif len(oligos) == 3:\n",
        "  plt_title = ['Forward primer', 'Reverse primer', 'Probe']\n",
        "  \n",
        "title_list = []\n",
        "for i in range(n_primer_pairs):\n",
        "    for oligo in plt_title:\n",
        "      title_list.append(f\"{oligo} {i}\")\n",
        "from plotly.subplots import make_subplots\n",
        "import plotly.graph_objects as go\n",
        "\n",
        "name = 'vgsc'\n",
        "oligos, _ = AgamPrimer.return_oligo_list(assay_type)\n",
        "\n",
        "hover_template = \"<br>\".join([\"Base / Position: %{customdata[4]}\",\n",
        "                                            \"Total Alternate freq: %{y}\",\n",
        "                                            \"A_freq: %{customdata[0]}\",\n",
        "                                            \"C_freq: %{customdata[1]}\",\n",
        "                                            \"G_freq: %{customdata[2]}\",\n",
        "                                            \"T_freq: %{customdata[3]}\"])\n",
        "                             \n",
        "fig = make_subplots(rows=n_primer_pairs, cols=len(oligos), subplot_titles=title_list, horizontal_spacing = 0.03, vertical_spacing=0.05)\n",
        "for idx, oligo in enumerate(oligos):\n",
        "  idx = idx+1\n",
        "  for i in range(n_primer_pairs):\n",
        "    row_i = i+1\n",
        "    fig.add_trace(go.Scatter(x=res_dict[i][oligo]['base_pos'], \n",
        "                             y=res_dict[i][oligo]['alt_frequency'], customdata=res_dict[i][oligo][['A_freq', 'C_freq', 'G_freq', 'T_freq', 'base_pos']], \n",
        "                             hovertemplate=hover_template,\n",
        "                             mode='markers', marker=dict(size=12, line=dict(width=2, color='black')), line=dict(color='dodgerblue'), marker_symbol='circle'), row=row_i, col=idx)\n",
        "    fig.add_annotation(row=row_i, col=idx, x=res_dict[i][oligo]['base_pos'][0], y=0.8,\n",
        "            text=\"5'\",\n",
        "            showarrow=False)\n",
        "    fig.add_annotation(row=row_i, col=idx, x=res_dict[i][oligo]['base_pos'].to_numpy()[-1], y=0.8,\n",
        "            text=\"3'\",\n",
        "            showarrow=False)\n",
        "    \n",
        "    fig.update_xaxes(row=row_i, col=idx, tickmode = 'array', tickvals = res_dict[i][oligo]['base_pos'], ticktext=res_dict[i][oligo]['base'], tickangle=0, mirror=True)\n",
        "    if idx > 1:\n",
        "      fig.update_yaxes(row=row_i, col=idx, range=[0,1], tickvals=np.arange(0, 1, 0.2), showticklabels=False, mirror=True)\n",
        "    else:\n",
        "      fig.update_yaxes(row=row_i, col=idx, tickvals=np.arange(0, 1, 0.2), range=[0,1], mirror=True)\n",
        "\n",
        "    if ((row_i % 2) == 0) & (idx == 1):\n",
        "      fig.update_yaxes(row=row_i, col=idx, title='Alternate allele frequency')\n",
        "\n",
        "if any(item in assay_type for item in ['gDNA']):\n",
        "  title_text = f\"{name} primer pairs | {sample_set} | target {target_loc} bp\"\n",
        "elif assay_type == 'probe':\n",
        "  title_text = f\"{name} probe | {sample_set} | target {target_loc} bp\"\n",
        "elif assay_type == 'qPCR primers':\n",
        "    title_text = f\"{name} primer pairs | {sample_set} | target {transcript}\"\n",
        "\n",
        "#fig.update_traces(customdata=customdata, hovertemplate=hovertemplate)\n",
        "fig.update_layout(height=1600, width=1600,\n",
        "                    title_text=title_text, title_x = 0.5, template=\"simple_white\", showlegend=False)\n",
        "\n",
        "fig.show()"
      ],
      "metadata": {
        "id": "NFNIXZHDZHPJ",
        "outputId": "dc9c5ca1-e9ca-4488-c8e0-830a1737e9c4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "id": "NFNIXZHDZHPJ",
      "execution_count": 276,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<html>\n",
              "<head><meta charset=\"utf-8\" /></head>\n",
              "<body>\n",
              "    <div>            <script src=\"https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_SVG\"></script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script>                <script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script>\n",
              "        <script src=\"https://cdn.plot.ly/plotly-2.8.3.min.js\"></script>                <div id=\"42917f08-771c-4bdb-8260-f0ce5202d774\" class=\"plotly-graph-div\" style=\"height:1600px; width:1600px;\"></div>            <script type=\"text/javascript\">                                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById(\"42917f08-771c-4bdb-8260-f0ce5202d774\")) {                    Plotly.newPlot(                        \"42917f08-771c-4bdb-8260-f0ce5202d774\",                        [{\"customdata\":[[1.0,0.0,0.0,0.0,\"A_28547103.0\"],[0.0,0.0,1.0,0.0,\"G_28547104.0\"],[0.0,0.0,0.0,1.0,\"T_28547105.0\"],[0.0,1.0,0.0,0.0,\"C_28547106.0\"],[0.0,1.0,0.0,0.0,\"C_28547107.0\"],[0.0,0.0,1.0,0.0,\"G_28547108.0\"],[0.0,0.0,1.0,0.0,\"G_28547109.0\"],[0.0,0.0,0.0,1.0,\"T_28547110.0\"],[0.0,1.0,0.0,0.0,\"C_28547111.0\"],[0.0,1.0,0.0,0.0,\"C_28547112.0\"],[0.0,0.0,1.0,0.0,\"G_28547113.0\"],[0.0,0.0,0.0,1.0,\"T_28547114.0\"],[1.0,0.0,0.0,0.0,\"A_28547115.0\"],[0.0,1.0,0.0,0.0,\"C_28547116.0\"],[1.0,0.0,0.0,0.0,\"A_28547117.0\"],[0.0,0.0,0.0,1.0,\"T_28547118.0\"],[0.0,0.0,0.0,1.0,\"T_28547119.0\"],[0.0,0.0,0.0,1.0,\"T_28547120.0\"],[0.0,1.0,0.0,0.0,\"C_28547121.0\"],[0.0,0.0,1.0,0.0,\"G_28547122.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"A_28547103.0\",\"G_28547104.0\",\"T_28547105.0\",\"C_28547106.0\",\"C_28547107.0\",\"G_28547108.0\",\"G_28547109.0\",\"T_28547110.0\",\"C_28547111.0\",\"C_28547112.0\",\"G_28547113.0\",\"T_28547114.0\",\"A_28547115.0\",\"C_28547116.0\",\"A_28547117.0\",\"T_28547118.0\",\"T_28547119.0\",\"T_28547120.0\",\"C_28547121.0\",\"G_28547122.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x\",\"yaxis\":\"y\"},{\"customdata\":[[0.0,1.0,0.0,0.0,\"C_28545668.0\"],[1.0,0.0,0.0,0.0,\"A_28545669.0\"],[0.0,1.0,0.0,0.0,\"C_28545670.0\"],[0.0,0.0,0.0,1.0,\"T_28545671.0\"],[0.0,0.0,0.0,1.0,\"T_28545672.0\"],[0.0,0.0,1.0,0.0,\"G_28545673.0\"],[0.0,0.0,0.0,1.0,\"T_28545674.0\"],[1.0,0.0,0.0,0.0,\"A_28545675.0\"],[0.0,0.99,0.0,0.01,\"C_28545676.0\"],[0.0,0.035,0.965,0.0,\"G_28545677.0\"],[0.0,0.0,1.0,0.0,\"G_28545678.0\"],[0.0,0.99,0.0,0.01,\"C_28545679.0\"],[0.0,0.0,0.0,1.0,\"T_28545680.0\"],[0.0,1.0,0.0,0.0,\"C_28545681.0\"],[0.0,1.0,0.0,0.0,\"C_28545682.0\"],[1.0,0.0,0.0,0.0,\"A_28545683.0\"],[0.0,1.0,0.0,0.0,\"C_28545684.0\"],[0.0,1.0,0.0,0.0,\"C_28545685.0\"],[0.0,0.0,0.0,1.0,\"T_28545686.0\"],[0.0,0.0,0.0,1.0,\"T_28545687.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"C_28545668.0\",\"A_28545669.0\",\"C_28545670.0\",\"T_28545671.0\",\"T_28545672.0\",\"G_28545673.0\",\"T_28545674.0\",\"A_28545675.0\",\"C_28545676.0\",\"G_28545677.0\",\"G_28545678.0\",\"C_28545679.0\",\"T_28545680.0\",\"C_28545681.0\",\"C_28545682.0\",\"A_28545683.0\",\"C_28545684.0\",\"C_28545685.0\",\"T_28545686.0\",\"T_28545687.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.01,0.035,0.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x3\",\"yaxis\":\"y3\"},{\"customdata\":[[1.0,0.0,0.0,0.0,\"A_28546080.0\"],[0.0,0.0,1.0,0.0,\"G_28546081.0\"],[0.0,1.0,0.0,0.0,\"C_28546082.0\"],[0.0,0.0,0.0,1.0,\"T_28546083.0\"],[0.0,0.0,0.0,1.0,\"T_28546084.0\"],[1.0,0.0,0.0,0.0,\"A_28546085.0\"],[0.0,0.0,0.0,1.0,\"T_28546086.0\"],[0.0,1.0,0.0,0.0,\"C_28546087.0\"],[0.0,0.0,0.85,0.15,\"G_28546088.0\"],[0.0,0.0,1.0,0.0,\"G_28546089.0\"],[0.0,0.0,0.0,1.0,\"T_28546090.0\"],[0.0,1.0,0.0,0.0,\"C_28546091.0\"],[1.0,0.0,0.0,0.0,\"A_28546092.0\"],[0.0,0.0,0.0,1.0,\"T_28546093.0\"],[0.0,0.0,1.0,0.0,\"G_28546094.0\"],[1.0,0.0,0.0,0.0,\"A_28546095.0\"],[0.0,0.0,1.0,0.0,\"G_28546096.0\"],[0.0,1.0,0.0,0.0,\"C_28546173.0\"],[0.0,0.0,1.0,0.0,\"G_28546174.0\"],[0.0,0.0,0.0,1.0,\"T_28546175.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"A_28546080.0\",\"G_28546081.0\",\"C_28546082.0\",\"T_28546083.0\",\"T_28546084.0\",\"A_28546085.0\",\"T_28546086.0\",\"C_28546087.0\",\"G_28546088.0\",\"G_28546089.0\",\"T_28546090.0\",\"C_28546091.0\",\"A_28546092.0\",\"T_28546093.0\",\"G_28546094.0\",\"A_28546095.0\",\"G_28546096.0\",\"C_28546173.0\",\"G_28546174.0\",\"T_28546175.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.15,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x5\",\"yaxis\":\"y5\"},{\"customdata\":[[1.0,0.0,0.0,0.0,\"A_28545761.0\"],[1.0,0.0,0.0,0.0,\"A_28545762.0\"],[0.0,1.0,0.0,0.0,\"C_28545763.0\"],[1.0,0.0,0.0,0.0,\"A_28545764.0\"],[1.0,0.0,0.0,0.0,\"A_28545765.0\"],[0.0,1.0,0.0,0.0,\"C_28545766.0\"],[0.035,0.0,0.0,0.965,\"T_28545767.0\"],[0.0,1.0,0.0,0.0,\"C_28545768.0\"],[0.0,0.0,1.0,0.0,\"G_28545769.0\"],[1.0,0.0,0.0,0.0,\"A_28545770.0\"],[0.0,1.0,0.0,0.0,\"C_28545771.0\"],[0.0,1.0,0.0,0.0,\"C_28545842.0\"],[1.0,0.0,0.0,0.0,\"A_28545843.0\"],[0.0,0.0,1.0,0.0,\"G_28545844.0\"],[0.0,1.0,0.0,0.0,\"C_28545845.0\"],[0.0,0.0,1.0,0.0,\"G_28545846.0\"],[0.0,0.0,0.0,1.0,\"T_28545847.0\"],[0.0,0.0,0.0,1.0,\"T_28545848.0\"],[0.0,0.0,0.0,1.0,\"T_28545849.0\"],[0.0,0.0,1.0,0.0,\"G_28545850.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"A_28545761.0\",\"A_28545762.0\",\"C_28545763.0\",\"A_28545764.0\",\"A_28545765.0\",\"C_28545766.0\",\"T_28545767.0\",\"C_28545768.0\",\"G_28545769.0\",\"A_28545770.0\",\"C_28545771.0\",\"C_28545842.0\",\"A_28545843.0\",\"G_28545844.0\",\"C_28545845.0\",\"G_28545846.0\",\"T_28545847.0\",\"T_28545848.0\",\"T_28545849.0\",\"G_28545850.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.035,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x7\",\"yaxis\":\"y7\"},{\"customdata\":[[0.0,0.0,0.0,1.0,\"T_28547184.0\"],[0.0,0.0,0.0,1.0,\"T_28547185.0\"],[0.0,0.0,1.0,0.0,\"G_28547186.0\"],[0.0,0.995,0.0,0.005,\"C_28547187.0\"],[0.0,1.0,0.0,0.0,\"C_28547188.0\"],[0.0,0.0,1.0,0.0,\"G_28547189.0\"],[0.0,0.0,1.0,0.0,\"G_28547190.0\"],[0.0,0.0,0.0,1.0,\"T_28547191.0\"],[0.0,0.0,0.0,1.0,\"T_28547192.0\"],[0.0,0.0,0.0,1.0,\"T_28547193.0\"],[0.0,0.0,1.0,0.0,\"G_28547194.0\"],[1.0,0.0,0.0,0.0,\"A_28547195.0\"],[0.0,0.0,1.0,0.0,\"G_28547196.0\"],[0.0,1.0,0.0,0.0,\"C_28547419.0\"],[0.0,0.0,0.0,1.0,\"T_28547420.0\"],[0.0,1.0,0.0,0.0,\"C_28547421.0\"],[0.0,0.0,0.0,1.0,\"T_28547422.0\"],[0.0,0.0,0.0,1.0,\"T_28547423.0\"],[0.0,0.0,0.0,1.0,\"T_28547424.0\"],[0.0,1.0,0.0,0.0,\"C_28547425.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"T_28547184.0\",\"T_28547185.0\",\"G_28547186.0\",\"C_28547187.0\",\"C_28547188.0\",\"G_28547189.0\",\"G_28547190.0\",\"T_28547191.0\",\"T_28547192.0\",\"T_28547193.0\",\"G_28547194.0\",\"A_28547195.0\",\"G_28547196.0\",\"C_28547419.0\",\"T_28547420.0\",\"C_28547421.0\",\"T_28547422.0\",\"T_28547423.0\",\"T_28547424.0\",\"C_28547425.0\"],\"y\":[0.0,0.0,0.0,0.005,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x9\",\"yaxis\":\"y9\"},{\"customdata\":[[0.0,1.0,0.0,0.0,\"C_28547116.0\"],[1.0,0.0,0.0,0.0,\"A_28547117.0\"],[0.0,0.0,0.0,1.0,\"T_28547118.0\"],[0.0,0.0,0.0,1.0,\"T_28547119.0\"],[0.0,0.0,0.0,1.0,\"T_28547120.0\"],[0.0,1.0,0.0,0.0,\"C_28547121.0\"],[0.0,0.0,1.0,0.0,\"G_28547122.0\"],[0.0,0.0,1.0,0.0,\"G_28547123.0\"],[0.0,0.0,0.0,1.0,\"T_28547124.0\"],[0.0,0.005,0.0,0.995,\"T_28547125.0\"],[0.0,1.0,0.0,0.0,\"C_28547126.0\"],[0.0,1.0,0.0,0.0,\"C_28547127.0\"],[0.0,0.0,1.0,0.0,\"G_28547128.0\"],[0.0,1.0,0.0,0.0,\"C_28547129.0\"],[0.0,0.0,0.0,1.0,\"T_28547130.0\"],[0.0,0.0,0.0,1.0,\"T_28547131.0\"],[0.0,0.0,1.0,0.0,\"G_28547132.0\"],[1.0,0.0,0.0,0.0,\"A_28547133.0\"],[0.145,0.0,0.855,0.0,\"G_28547134.0\"],[0.0,1.0,0.0,0.0,\"C_28547135.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"C_28547116.0\",\"A_28547117.0\",\"T_28547118.0\",\"T_28547119.0\",\"T_28547120.0\",\"C_28547121.0\",\"G_28547122.0\",\"G_28547123.0\",\"T_28547124.0\",\"T_28547125.0\",\"C_28547126.0\",\"C_28547127.0\",\"G_28547128.0\",\"C_28547129.0\",\"T_28547130.0\",\"T_28547131.0\",\"G_28547132.0\",\"A_28547133.0\",\"G_28547134.0\",\"C_28547135.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.005,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.145,0.0],\"type\":\"scatter\",\"xaxis\":\"x11\",\"yaxis\":\"y11\"},{\"customdata\":[[0.235,0.765,0.0,0.0,\"C_28547428.0\"],[0.0,1.0,0.0,0.0,\"A_28547427.0\"],[0.0,0.0,1.0,0.0,\"C_28547426.0\"],[0.0,0.0,1.0,0.0,\"G_28547425.0\"],[0.0,0.0,0.0,1.0,\"A_28547424.0\"],[0.0,0.0,0.0,1.0,\"A_28547423.0\"],[0.0,0.0,0.0,1.0,\"A_28547422.0\"],[0.0,0.0,1.0,0.0,\"G_28547421.0\"],[1.0,0.0,0.0,0.0,\"A_28547420.0\"],[0.0,0.0,1.0,0.0,\"G_28547419.0\"],[0.0,1.0,0.0,0.0,\"C_28547196.0\"],[0.0,0.0,0.0,1.0,\"T_28547195.0\"],[0.0,1.0,0.0,0.0,\"C_28547194.0\"],[0.0,0.0,0.0,1.0,\"A_28547193.0\"],[0.0,0.0,0.0,1.0,\"A_28547192.0\"],[0.0,0.0,0.0,1.0,\"A_28547191.0\"],[0.0,1.0,0.0,0.0,\"C_28547190.0\"],[0.0,0.0,1.0,0.0,\"C_28547189.0\"],[0.0,0.0,0.0,1.0,\"G_28547188.0\"],[0.0,0.005,0.995,0.0,\"G_28547187.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"C_28547428.0\",\"A_28547427.0\",\"C_28547426.0\",\"G_28547425.0\",\"A_28547424.0\",\"A_28547423.0\",\"A_28547422.0\",\"G_28547421.0\",\"A_28547420.0\",\"G_28547419.0\",\"C_28547196.0\",\"T_28547195.0\",\"C_28547194.0\",\"A_28547193.0\",\"A_28547192.0\",\"A_28547191.0\",\"C_28547190.0\",\"C_28547189.0\",\"G_28547188.0\",\"G_28547187.0\"],\"y\":[0.235,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.005],\"type\":\"scatter\",\"xaxis\":\"x2\",\"yaxis\":\"y2\"},{\"customdata\":[[1.0,0.0,0.0,0.0,\"C_28545850.0\"],[1.0,0.0,0.0,0.0,\"A_28545849.0\"],[0.0,1.0,0.0,0.0,\"A_28545848.0\"],[1.0,0.0,0.0,0.0,\"A_28545847.0\"],[1.0,0.0,0.0,0.0,\"C_28545846.0\"],[0.0,1.0,0.0,0.0,\"G_28545845.0\"],[0.0,0.0,0.0,1.0,\"C_28545844.0\"],[0.0,1.0,0.0,0.0,\"T_28545843.0\"],[0.0,0.0,1.0,0.0,\"G_28545842.0\"],[1.0,0.0,0.0,0.0,\"G_28545771.0\"],[0.0,1.0,0.0,0.0,\"T_28545770.0\"],[0.0,1.0,0.0,0.0,\"C_28545769.0\"],[1.0,0.0,0.0,0.0,\"G_28545768.0\"],[0.035,0.0,0.965,0.0,\"A_28545767.0\"],[0.0,1.0,0.0,0.0,\"G_28545766.0\"],[0.0,0.0,1.0,0.0,\"T_28545765.0\"],[0.0,0.0,0.0,1.0,\"T_28545764.0\"],[0.0,0.0,0.0,1.0,\"G_28545763.0\"],[0.0,0.0,0.0,1.0,\"T_28545762.0\"],[0.0,0.0,1.0,0.0,\"T_28545761.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"C_28545850.0\",\"A_28545849.0\",\"A_28545848.0\",\"A_28545847.0\",\"C_28545846.0\",\"G_28545845.0\",\"C_28545844.0\",\"T_28545843.0\",\"G_28545842.0\",\"G_28545771.0\",\"T_28545770.0\",\"C_28545769.0\",\"G_28545768.0\",\"A_28545767.0\",\"G_28545766.0\",\"T_28545765.0\",\"T_28545764.0\",\"G_28545763.0\",\"T_28545762.0\",\"T_28545761.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.035,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x4\",\"yaxis\":\"y4\"},{\"customdata\":[[1.0,0.0,0.0,0.0,\"G_28546217.0\"],[1.0,0.0,0.0,0.0,\"G_28546216.0\"],[0.0,1.0,0.0,0.0,\"T_28546215.0\"],[0.0,0.0,1.0,0.0,\"G_28546214.0\"],[0.14,0.025,0.0,0.835,\"G_28546213.0\"],[0.0,0.0,0.0,1.0,\"A_28546212.0\"],[0.0,0.0,1.0,0.0,\"T_28546211.0\"],[0.0,0.0,1.0,0.0,\"C_28546210.0\"],[0.02,0.0,0.0,0.98,\"C_28546209.0\"],[0.0,0.0,0.0,1.0,\"C_28546208.0\"],[0.015,0.0,0.985,0.0,\"A_28546207.0\"],[0.0,0.0,1.0,0.0,\"A_28546206.0\"],[0.0,0.0,1.0,0.0,\"C_28546205.0\"],[1.0,0.0,0.0,0.0,\"C_28546204.0\"],[0.0,0.0,0.0,1.0,\"A_28546203.0\"],[0.0,1.0,0.0,0.0,\"T_28546202.0\"],[0.0,1.0,0.0,0.0,\"C_28546201.0\"],[1.0,0.0,0.0,0.0,\"G_28546200.0\"],[0.0,1.0,0.0,0.0,\"T_28546199.0\"],[0.0,1.0,0.0,0.0,\"T_28546198.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"G_28546217.0\",\"G_28546216.0\",\"T_28546215.0\",\"G_28546214.0\",\"G_28546213.0\",\"A_28546212.0\",\"T_28546211.0\",\"C_28546210.0\",\"C_28546209.0\",\"C_28546208.0\",\"A_28546207.0\",\"A_28546206.0\",\"C_28546205.0\",\"C_28546204.0\",\"A_28546203.0\",\"T_28546202.0\",\"C_28546201.0\",\"G_28546200.0\",\"T_28546199.0\",\"T_28546198.0\"],\"y\":[0.0,0.0,0.0,0.0,0.86,0.0,0.0,0.0,0.02,0.0,0.015,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x6\",\"yaxis\":\"y6\"},{\"customdata\":[[0.0,1.0,0.0,0.0,\"T_28545905.0\"],[0.0,1.0,0.0,0.0,\"A_28545904.0\"],[0.0,0.0,1.0,0.0,\"A_28545903.0\"],[0.0,0.0,1.0,0.0,\"G_28545902.0\"],[0.0,0.0,1.0,0.0,\"A_28545901.0\"],[0.0,1.0,0.0,0.0,\"A_28545900.0\"],[0.0,0.0,0.0,1.0,\"C_28545899.0\"],[0.0,0.0,1.0,0.0,\"G_28545898.0\"],[1.0,0.0,0.0,0.0,\"T_28545897.0\"],[1.0,0.0,0.0,0.0,\"G_28545896.0\"],[0.0,1.0,0.0,0.0,\"T_28545895.0\"],[1.0,0.0,0.0,0.0,\"T_28545894.0\"],[0.0,1.0,0.0,0.0,\"C_28545893.0\"],[0.0,0.0,1.0,0.0,\"A_28545892.0\"],[0.0,0.0,0.0,1.0,\"G_28545891.0\"],[0.03,0.0,0.0,0.97,\"C_28545890.0\"],[0.0,1.0,0.0,0.0,\"C_28545889.0\"],[0.0,0.0,0.0,1.0,\"C_28545888.0\"],[0.005,0.0,0.0,0.995,\"G_28545887.0\"],[1.0,0.0,0.0,0.0,\"G_28545886.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"T_28545905.0\",\"A_28545904.0\",\"A_28545903.0\",\"G_28545902.0\",\"A_28545901.0\",\"A_28545900.0\",\"C_28545899.0\",\"G_28545898.0\",\"T_28545897.0\",\"G_28545896.0\",\"T_28545895.0\",\"T_28545894.0\",\"C_28545893.0\",\"A_28545892.0\",\"G_28545891.0\",\"C_28545890.0\",\"C_28545889.0\",\"C_28545888.0\",\"G_28545887.0\",\"G_28545886.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.03,0.0,0.0,0.005,0.0],\"type\":\"scatter\",\"xaxis\":\"x8\",\"yaxis\":\"y8\"},{\"customdata\":[[0.0,1.0,0.0,0.0,\"A_28547471.0\"],[0.0,0.0,0.0,1.0,\"C_28547470.0\"],[0.0,1.0,0.0,0.0,\"A_28547469.0\"],[1.0,0.0,0.0,0.0,\"A_28547468.0\"],[0.0,1.0,0.0,0.0,\"G_28547467.0\"],[0.0,0.0,0.0,1.0,\"A_28547466.0\"],[0.0,0.0,1.0,0.0,\"T_28547465.0\"],[0.0,1.0,0.0,0.0,\"C_28547464.0\"],[0.0,1.0,0.0,0.0,\"G_28547463.0\"],[0.0,1.0,0.0,0.0,\"T_28547462.0\"],[1.0,0.0,0.0,0.0,\"G_28547461.0\"],[0.0,1.0,0.0,0.0,\"G_28547460.0\"],[0.0,0.0,1.0,0.0,\"G_28547459.0\"],[1.0,0.0,0.0,0.0,\"C_28547458.0\"],[0.0,0.0,0.0,1.0,\"A_28547457.0\"],[0.0,1.0,0.0,0.0,\"G_28547456.0\"],[0.0,0.0,0.0,1.0,\"T_28547455.0\"],[0.0,0.0,0.0,1.0,\"G_28547454.0\"],[0.0,0.0,1.0,0.0,\"A_28547453.0\"],[0.0,0.0,0.0,1.0,\"G_28547452.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"A_28547471.0\",\"C_28547470.0\",\"A_28547469.0\",\"A_28547468.0\",\"G_28547467.0\",\"A_28547466.0\",\"T_28547465.0\",\"C_28547464.0\",\"G_28547463.0\",\"T_28547462.0\",\"G_28547461.0\",\"G_28547460.0\",\"G_28547459.0\",\"C_28547458.0\",\"A_28547457.0\",\"G_28547456.0\",\"T_28547455.0\",\"G_28547454.0\",\"A_28547453.0\",\"G_28547452.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x10\",\"yaxis\":\"y10\"},{\"customdata\":[[0.0,0.0,0.0,1.0,\"G_28547425.0\"],[0.0,0.0,0.0,1.0,\"A_28547424.0\"],[0.0,0.0,1.0,0.0,\"A_28547423.0\"],[0.0,1.0,0.0,0.0,\"A_28547422.0\"],[0.0,1.0,0.0,0.0,\"G_28547421.0\"],[0.0,0.0,1.0,0.0,\"A_28547420.0\"],[0.0,0.0,1.0,0.0,\"G_28547419.0\"],[0.0,0.0,0.0,1.0,\"C_28547196.0\"],[0.0,0.0,0.0,1.0,\"T_28547195.0\"],[0.0,0.0,0.0,1.0,\"C_28547194.0\"],[0.0,0.0,1.0,0.0,\"A_28547193.0\"],[1.0,0.0,0.0,0.0,\"A_28547192.0\"],[0.0,0.0,1.0,0.0,\"A_28547191.0\"],[0.0,1.0,0.0,0.0,\"C_28547190.0\"],[0.0,0.0,0.0,1.0,\"C_28547189.0\"],[0.0,1.0,0.0,0.0,\"G_28547188.0\"],[0.0,0.005,0.0,0.995,\"G_28547187.0\"],[0.0,0.0,0.0,1.0,\"C_28547186.0\"],[0.0,0.0,0.0,1.0,\"A_28547185.0\"],[0.0,1.0,0.0,0.0,\"A_28547184.0\"]],\"hovertemplate\":\"Base / Position: %{customdata[4]}<br>Total Alternate freq: %{y}<br>A_freq: %{customdata[0]}<br>C_freq: %{customdata[1]}<br>G_freq: %{customdata[2]}<br>T_freq: %{customdata[3]}\",\"line\":{\"color\":\"dodgerblue\"},\"marker\":{\"line\":{\"color\":\"black\",\"width\":2},\"size\":12,\"symbol\":\"circle\"},\"mode\":\"markers\",\"x\":[\"G_28547425.0\",\"A_28547424.0\",\"A_28547423.0\",\"A_28547422.0\",\"G_28547421.0\",\"A_28547420.0\",\"G_28547419.0\",\"C_28547196.0\",\"T_28547195.0\",\"C_28547194.0\",\"A_28547193.0\",\"A_28547192.0\",\"A_28547191.0\",\"C_28547190.0\",\"C_28547189.0\",\"G_28547188.0\",\"G_28547187.0\",\"C_28547186.0\",\"A_28547185.0\",\"A_28547184.0\"],\"y\":[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.005,0.0,0.0,0.0],\"type\":\"scatter\",\"xaxis\":\"x12\",\"yaxis\":\"y12\"}],                        {\"template\":{\"data\":{\"barpolar\":[{\"marker\":{\"line\":{\"color\":\"white\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"barpolar\"}],\"bar\":[{\"error_x\":{\"color\":\"rgb(36,36,36)\"},\"error_y\":{\"color\":\"rgb(36,36,36)\"},\"marker\":{\"line\":{\"color\":\"white\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"bar\"}],\"carpet\":[{\"aaxis\":{\"endlinecolor\":\"rgb(36,36,36)\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"rgb(36,36,36)\"},\"baxis\":{\"endlinecolor\":\"rgb(36,36,36)\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"rgb(36,36,36)\"},\"type\":\"carpet\"}],\"choropleth\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"type\":\"choropleth\"}],\"contourcarpet\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"type\":\"contourcarpet\"}],\"contour\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"colorscale\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]],\"type\":\"contour\"}],\"heatmapgl\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"colorscale\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]],\"type\":\"heatmapgl\"}],\"heatmap\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"colorscale\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]],\"type\":\"heatmap\"}],\"histogram2dcontour\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"colorscale\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]],\"type\":\"histogram2dcontour\"}],\"histogram2d\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"colorscale\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]],\"type\":\"histogram2d\"}],\"histogram\":[{\"marker\":{\"line\":{\"color\":\"white\",\"width\":0.6}},\"type\":\"histogram\"}],\"mesh3d\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"type\":\"mesh3d\"}],\"parcoords\":[{\"line\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"parcoords\"}],\"pie\":[{\"automargin\":true,\"type\":\"pie\"}],\"scatter3d\":[{\"line\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scatter3d\"}],\"scattercarpet\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scattercarpet\"}],\"scattergeo\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scattergeo\"}],\"scattergl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scattergl\"}],\"scattermapbox\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scattermapbox\"}],\"scatterpolargl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scatterpolargl\"}],\"scatterpolar\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scatterpolar\"}],\"scatter\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scatter\"}],\"scatterternary\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"type\":\"scatterternary\"}],\"surface\":[{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"},\"colorscale\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]],\"type\":\"surface\"}],\"table\":[{\"cells\":{\"fill\":{\"color\":\"rgb(237,237,237)\"},\"line\":{\"color\":\"white\"}},\"header\":{\"fill\":{\"color\":\"rgb(217,217,217)\"},\"line\":{\"color\":\"white\"}},\"type\":\"table\"}]},\"layout\":{\"annotationdefaults\":{\"arrowhead\":0,\"arrowwidth\":1},\"autotypenumbers\":\"strict\",\"coloraxis\":{\"colorbar\":{\"outlinewidth\":1,\"tickcolor\":\"rgb(36,36,36)\",\"ticks\":\"outside\"}},\"colorscale\":{\"diverging\":[[0.0,\"rgb(103,0,31)\"],[0.1,\"rgb(178,24,43)\"],[0.2,\"rgb(214,96,77)\"],[0.3,\"rgb(244,165,130)\"],[0.4,\"rgb(253,219,199)\"],[0.5,\"rgb(247,247,247)\"],[0.6,\"rgb(209,229,240)\"],[0.7,\"rgb(146,197,222)\"],[0.8,\"rgb(67,147,195)\"],[0.9,\"rgb(33,102,172)\"],[1.0,\"rgb(5,48,97)\"]],\"sequential\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]],\"sequentialminus\":[[0.0,\"#440154\"],[0.1111111111111111,\"#482878\"],[0.2222222222222222,\"#3e4989\"],[0.3333333333333333,\"#31688e\"],[0.4444444444444444,\"#26828e\"],[0.5555555555555556,\"#1f9e89\"],[0.6666666666666666,\"#35b779\"],[0.7777777777777778,\"#6ece58\"],[0.8888888888888888,\"#b5de2b\"],[1.0,\"#fde725\"]]},\"colorway\":[\"#1F77B4\",\"#FF7F0E\",\"#2CA02C\",\"#D62728\",\"#9467BD\",\"#8C564B\",\"#E377C2\",\"#7F7F7F\",\"#BCBD22\",\"#17BECF\"],\"font\":{\"color\":\"rgb(36,36,36)\"},\"geo\":{\"bgcolor\":\"white\",\"lakecolor\":\"white\",\"landcolor\":\"white\",\"showlakes\":true,\"showland\":true,\"subunitcolor\":\"white\"},\"hoverlabel\":{\"align\":\"left\"},\"hovermode\":\"closest\",\"mapbox\":{\"style\":\"light\"},\"paper_bgcolor\":\"white\",\"plot_bgcolor\":\"white\",\"polar\":{\"angularaxis\":{\"gridcolor\":\"rgb(232,232,232)\",\"linecolor\":\"rgb(36,36,36)\",\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\"},\"bgcolor\":\"white\",\"radialaxis\":{\"gridcolor\":\"rgb(232,232,232)\",\"linecolor\":\"rgb(36,36,36)\",\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\"}},\"scene\":{\"xaxis\":{\"backgroundcolor\":\"white\",\"gridcolor\":\"rgb(232,232,232)\",\"gridwidth\":2,\"linecolor\":\"rgb(36,36,36)\",\"showbackground\":true,\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\",\"zeroline\":false,\"zerolinecolor\":\"rgb(36,36,36)\"},\"yaxis\":{\"backgroundcolor\":\"white\",\"gridcolor\":\"rgb(232,232,232)\",\"gridwidth\":2,\"linecolor\":\"rgb(36,36,36)\",\"showbackground\":true,\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\",\"zeroline\":false,\"zerolinecolor\":\"rgb(36,36,36)\"},\"zaxis\":{\"backgroundcolor\":\"white\",\"gridcolor\":\"rgb(232,232,232)\",\"gridwidth\":2,\"linecolor\":\"rgb(36,36,36)\",\"showbackground\":true,\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\",\"zeroline\":false,\"zerolinecolor\":\"rgb(36,36,36)\"}},\"shapedefaults\":{\"fillcolor\":\"black\",\"line\":{\"width\":0},\"opacity\":0.3},\"ternary\":{\"aaxis\":{\"gridcolor\":\"rgb(232,232,232)\",\"linecolor\":\"rgb(36,36,36)\",\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\"},\"baxis\":{\"gridcolor\":\"rgb(232,232,232)\",\"linecolor\":\"rgb(36,36,36)\",\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\"},\"bgcolor\":\"white\",\"caxis\":{\"gridcolor\":\"rgb(232,232,232)\",\"linecolor\":\"rgb(36,36,36)\",\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\"}},\"title\":{\"x\":0.05},\"xaxis\":{\"automargin\":true,\"gridcolor\":\"rgb(232,232,232)\",\"linecolor\":\"rgb(36,36,36)\",\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\",\"title\":{\"standoff\":15},\"zeroline\":false,\"zerolinecolor\":\"rgb(36,36,36)\"},\"yaxis\":{\"automargin\":true,\"gridcolor\":\"rgb(232,232,232)\",\"linecolor\":\"rgb(36,36,36)\",\"showgrid\":false,\"showline\":true,\"ticks\":\"outside\",\"title\":{\"standoff\":15},\"zeroline\":false,\"zerolinecolor\":\"rgb(36,36,36)\"}}},\"xaxis\":{\"anchor\":\"y\",\"domain\":[0.0,0.485],\"tickmode\":\"array\",\"tickvals\":[\"A_28547103.0\",\"G_28547104.0\",\"T_28547105.0\",\"C_28547106.0\",\"C_28547107.0\",\"G_28547108.0\",\"G_28547109.0\",\"T_28547110.0\",\"C_28547111.0\",\"C_28547112.0\",\"G_28547113.0\",\"T_28547114.0\",\"A_28547115.0\",\"C_28547116.0\",\"A_28547117.0\",\"T_28547118.0\",\"T_28547119.0\",\"T_28547120.0\",\"C_28547121.0\",\"G_28547122.0\"],\"ticktext\":[\"A\",\"G\",\"T\",\"C\",\"C\",\"G\",\"G\",\"T\",\"C\",\"C\",\"G\",\"T\",\"A\",\"C\",\"A\",\"T\",\"T\",\"T\",\"C\",\"G\"],\"tickangle\":0,\"mirror\":true},\"yaxis\":{\"anchor\":\"x\",\"domain\":[0.875,1.0],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"range\":[0,1],\"mirror\":true},\"xaxis2\":{\"anchor\":\"y2\",\"domain\":[0.515,1.0],\"tickmode\":\"array\",\"tickvals\":[\"C_28547428.0\",\"A_28547427.0\",\"C_28547426.0\",\"G_28547425.0\",\"A_28547424.0\",\"A_28547423.0\",\"A_28547422.0\",\"G_28547421.0\",\"A_28547420.0\",\"G_28547419.0\",\"C_28547196.0\",\"T_28547195.0\",\"C_28547194.0\",\"A_28547193.0\",\"A_28547192.0\",\"A_28547191.0\",\"C_28547190.0\",\"C_28547189.0\",\"G_28547188.0\",\"G_28547187.0\"],\"ticktext\":[\"C\",\"A\",\"C\",\"G\",\"A\",\"A\",\"A\",\"G\",\"A\",\"G\",\"C\",\"T\",\"C\",\"A\",\"A\",\"A\",\"C\",\"C\",\"G\",\"G\"],\"tickangle\":0,\"mirror\":true},\"yaxis2\":{\"anchor\":\"x2\",\"domain\":[0.875,1.0],\"range\":[0,1],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"showticklabels\":false,\"mirror\":true},\"xaxis3\":{\"anchor\":\"y3\",\"domain\":[0.0,0.485],\"tickmode\":\"array\",\"tickvals\":[\"C_28545668.0\",\"A_28545669.0\",\"C_28545670.0\",\"T_28545671.0\",\"T_28545672.0\",\"G_28545673.0\",\"T_28545674.0\",\"A_28545675.0\",\"C_28545676.0\",\"G_28545677.0\",\"G_28545678.0\",\"C_28545679.0\",\"T_28545680.0\",\"C_28545681.0\",\"C_28545682.0\",\"A_28545683.0\",\"C_28545684.0\",\"C_28545685.0\",\"T_28545686.0\",\"T_28545687.0\"],\"ticktext\":[\"C\",\"A\",\"C\",\"T\",\"T\",\"G\",\"T\",\"A\",\"C\",\"G\",\"G\",\"C\",\"T\",\"C\",\"C\",\"A\",\"C\",\"C\",\"T\",\"T\"],\"tickangle\":0,\"mirror\":true},\"yaxis3\":{\"anchor\":\"x3\",\"domain\":[0.7,0.825],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"range\":[0,1],\"mirror\":true,\"title\":{\"text\":\"Alternate allele frequency\"}},\"xaxis4\":{\"anchor\":\"y4\",\"domain\":[0.515,1.0],\"tickmode\":\"array\",\"tickvals\":[\"C_28545850.0\",\"A_28545849.0\",\"A_28545848.0\",\"A_28545847.0\",\"C_28545846.0\",\"G_28545845.0\",\"C_28545844.0\",\"T_28545843.0\",\"G_28545842.0\",\"G_28545771.0\",\"T_28545770.0\",\"C_28545769.0\",\"G_28545768.0\",\"A_28545767.0\",\"G_28545766.0\",\"T_28545765.0\",\"T_28545764.0\",\"G_28545763.0\",\"T_28545762.0\",\"T_28545761.0\"],\"ticktext\":[\"C\",\"A\",\"A\",\"A\",\"C\",\"G\",\"C\",\"T\",\"G\",\"G\",\"T\",\"C\",\"G\",\"A\",\"G\",\"T\",\"T\",\"G\",\"T\",\"T\"],\"tickangle\":0,\"mirror\":true},\"yaxis4\":{\"anchor\":\"x4\",\"domain\":[0.7,0.825],\"range\":[0,1],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"showticklabels\":false,\"mirror\":true},\"xaxis5\":{\"anchor\":\"y5\",\"domain\":[0.0,0.485],\"tickmode\":\"array\",\"tickvals\":[\"A_28546080.0\",\"G_28546081.0\",\"C_28546082.0\",\"T_28546083.0\",\"T_28546084.0\",\"A_28546085.0\",\"T_28546086.0\",\"C_28546087.0\",\"G_28546088.0\",\"G_28546089.0\",\"T_28546090.0\",\"C_28546091.0\",\"A_28546092.0\",\"T_28546093.0\",\"G_28546094.0\",\"A_28546095.0\",\"G_28546096.0\",\"C_28546173.0\",\"G_28546174.0\",\"T_28546175.0\"],\"ticktext\":[\"A\",\"G\",\"C\",\"T\",\"T\",\"A\",\"T\",\"C\",\"G\",\"G\",\"T\",\"C\",\"A\",\"T\",\"G\",\"A\",\"G\",\"C\",\"G\",\"T\"],\"tickangle\":0,\"mirror\":true},\"yaxis5\":{\"anchor\":\"x5\",\"domain\":[0.525,0.65],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"range\":[0,1],\"mirror\":true},\"xaxis6\":{\"anchor\":\"y6\",\"domain\":[0.515,1.0],\"tickmode\":\"array\",\"tickvals\":[\"G_28546217.0\",\"G_28546216.0\",\"T_28546215.0\",\"G_28546214.0\",\"G_28546213.0\",\"A_28546212.0\",\"T_28546211.0\",\"C_28546210.0\",\"C_28546209.0\",\"C_28546208.0\",\"A_28546207.0\",\"A_28546206.0\",\"C_28546205.0\",\"C_28546204.0\",\"A_28546203.0\",\"T_28546202.0\",\"C_28546201.0\",\"G_28546200.0\",\"T_28546199.0\",\"T_28546198.0\"],\"ticktext\":[\"G\",\"G\",\"T\",\"G\",\"G\",\"A\",\"T\",\"C\",\"C\",\"C\",\"A\",\"A\",\"C\",\"C\",\"A\",\"T\",\"C\",\"G\",\"T\",\"T\"],\"tickangle\":0,\"mirror\":true},\"yaxis6\":{\"anchor\":\"x6\",\"domain\":[0.525,0.65],\"range\":[0,1],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"showticklabels\":false,\"mirror\":true},\"xaxis7\":{\"anchor\":\"y7\",\"domain\":[0.0,0.485],\"tickmode\":\"array\",\"tickvals\":[\"A_28545761.0\",\"A_28545762.0\",\"C_28545763.0\",\"A_28545764.0\",\"A_28545765.0\",\"C_28545766.0\",\"T_28545767.0\",\"C_28545768.0\",\"G_28545769.0\",\"A_28545770.0\",\"C_28545771.0\",\"C_28545842.0\",\"A_28545843.0\",\"G_28545844.0\",\"C_28545845.0\",\"G_28545846.0\",\"T_28545847.0\",\"T_28545848.0\",\"T_28545849.0\",\"G_28545850.0\"],\"ticktext\":[\"A\",\"A\",\"C\",\"A\",\"A\",\"C\",\"T\",\"C\",\"G\",\"A\",\"C\",\"C\",\"A\",\"G\",\"C\",\"G\",\"T\",\"T\",\"T\",\"G\"],\"tickangle\":0,\"mirror\":true},\"yaxis7\":{\"anchor\":\"x7\",\"domain\":[0.35,0.475],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"range\":[0,1],\"mirror\":true,\"title\":{\"text\":\"Alternate allele frequency\"}},\"xaxis8\":{\"anchor\":\"y8\",\"domain\":[0.515,1.0],\"tickmode\":\"array\",\"tickvals\":[\"T_28545905.0\",\"A_28545904.0\",\"A_28545903.0\",\"G_28545902.0\",\"A_28545901.0\",\"A_28545900.0\",\"C_28545899.0\",\"G_28545898.0\",\"T_28545897.0\",\"G_28545896.0\",\"T_28545895.0\",\"T_28545894.0\",\"C_28545893.0\",\"A_28545892.0\",\"G_28545891.0\",\"C_28545890.0\",\"C_28545889.0\",\"C_28545888.0\",\"G_28545887.0\",\"G_28545886.0\"],\"ticktext\":[\"T\",\"A\",\"A\",\"G\",\"A\",\"A\",\"C\",\"G\",\"T\",\"G\",\"T\",\"T\",\"C\",\"A\",\"G\",\"C\",\"C\",\"C\",\"G\",\"G\"],\"tickangle\":0,\"mirror\":true},\"yaxis8\":{\"anchor\":\"x8\",\"domain\":[0.35,0.475],\"range\":[0,1],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"showticklabels\":false,\"mirror\":true},\"xaxis9\":{\"anchor\":\"y9\",\"domain\":[0.0,0.485],\"tickmode\":\"array\",\"tickvals\":[\"T_28547184.0\",\"T_28547185.0\",\"G_28547186.0\",\"C_28547187.0\",\"C_28547188.0\",\"G_28547189.0\",\"G_28547190.0\",\"T_28547191.0\",\"T_28547192.0\",\"T_28547193.0\",\"G_28547194.0\",\"A_28547195.0\",\"G_28547196.0\",\"C_28547419.0\",\"T_28547420.0\",\"C_28547421.0\",\"T_28547422.0\",\"T_28547423.0\",\"T_28547424.0\",\"C_28547425.0\"],\"ticktext\":[\"T\",\"T\",\"G\",\"C\",\"C\",\"G\",\"G\",\"T\",\"T\",\"T\",\"G\",\"A\",\"G\",\"C\",\"T\",\"C\",\"T\",\"T\",\"T\",\"C\"],\"tickangle\":0,\"mirror\":true},\"yaxis9\":{\"anchor\":\"x9\",\"domain\":[0.175,0.3],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"range\":[0,1],\"mirror\":true},\"xaxis10\":{\"anchor\":\"y10\",\"domain\":[0.515,1.0],\"tickmode\":\"array\",\"tickvals\":[\"A_28547471.0\",\"C_28547470.0\",\"A_28547469.0\",\"A_28547468.0\",\"G_28547467.0\",\"A_28547466.0\",\"T_28547465.0\",\"C_28547464.0\",\"G_28547463.0\",\"T_28547462.0\",\"G_28547461.0\",\"G_28547460.0\",\"G_28547459.0\",\"C_28547458.0\",\"A_28547457.0\",\"G_28547456.0\",\"T_28547455.0\",\"G_28547454.0\",\"A_28547453.0\",\"G_28547452.0\"],\"ticktext\":[\"A\",\"C\",\"A\",\"A\",\"G\",\"A\",\"T\",\"C\",\"G\",\"T\",\"G\",\"G\",\"G\",\"C\",\"A\",\"G\",\"T\",\"G\",\"A\",\"G\"],\"tickangle\":0,\"mirror\":true},\"yaxis10\":{\"anchor\":\"x10\",\"domain\":[0.175,0.3],\"range\":[0,1],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"showticklabels\":false,\"mirror\":true},\"xaxis11\":{\"anchor\":\"y11\",\"domain\":[0.0,0.485],\"tickmode\":\"array\",\"tickvals\":[\"C_28547116.0\",\"A_28547117.0\",\"T_28547118.0\",\"T_28547119.0\",\"T_28547120.0\",\"C_28547121.0\",\"G_28547122.0\",\"G_28547123.0\",\"T_28547124.0\",\"T_28547125.0\",\"C_28547126.0\",\"C_28547127.0\",\"G_28547128.0\",\"C_28547129.0\",\"T_28547130.0\",\"T_28547131.0\",\"G_28547132.0\",\"A_28547133.0\",\"G_28547134.0\",\"C_28547135.0\"],\"ticktext\":[\"C\",\"A\",\"T\",\"T\",\"T\",\"C\",\"G\",\"G\",\"T\",\"T\",\"C\",\"C\",\"G\",\"C\",\"T\",\"T\",\"G\",\"A\",\"G\",\"C\"],\"tickangle\":0,\"mirror\":true},\"yaxis11\":{\"anchor\":\"x11\",\"domain\":[0.0,0.125],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"range\":[0,1],\"mirror\":true,\"title\":{\"text\":\"Alternate allele frequency\"}},\"xaxis12\":{\"anchor\":\"y12\",\"domain\":[0.515,1.0],\"tickmode\":\"array\",\"tickvals\":[\"G_28547425.0\",\"A_28547424.0\",\"A_28547423.0\",\"A_28547422.0\",\"G_28547421.0\",\"A_28547420.0\",\"G_28547419.0\",\"C_28547196.0\",\"T_28547195.0\",\"C_28547194.0\",\"A_28547193.0\",\"A_28547192.0\",\"A_28547191.0\",\"C_28547190.0\",\"C_28547189.0\",\"G_28547188.0\",\"G_28547187.0\",\"C_28547186.0\",\"A_28547185.0\",\"A_28547184.0\"],\"ticktext\":[\"G\",\"A\",\"A\",\"A\",\"G\",\"A\",\"G\",\"C\",\"T\",\"C\",\"A\",\"A\",\"A\",\"C\",\"C\",\"G\",\"G\",\"C\",\"A\",\"A\"],\"tickangle\":0,\"mirror\":true},\"yaxis12\":{\"anchor\":\"x12\",\"domain\":[0.0,0.125],\"range\":[0,1],\"tickvals\":[0.0,0.2,0.4,0.6000000000000001,0.8],\"showticklabels\":false,\"mirror\":true},\"annotations\":[{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Forward primer 0\",\"x\":0.2425,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":1.0,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Reverse primer 0\",\"x\":0.7575000000000001,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":1.0,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Forward primer 1\",\"x\":0.2425,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.825,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Reverse primer 1\",\"x\":0.7575000000000001,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.825,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Forward primer 2\",\"x\":0.2425,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.65,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Reverse primer 2\",\"x\":0.7575000000000001,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.65,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Forward primer 3\",\"x\":0.2425,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.475,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Reverse primer 3\",\"x\":0.7575000000000001,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.475,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Forward primer 4\",\"x\":0.2425,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.3,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Reverse primer 4\",\"x\":0.7575000000000001,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.3,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Forward primer 5\",\"x\":0.2425,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.125,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"font\":{\"size\":16},\"showarrow\":false,\"text\":\"Reverse primer 5\",\"x\":0.7575000000000001,\"xanchor\":\"center\",\"xref\":\"paper\",\"y\":0.125,\"yanchor\":\"bottom\",\"yref\":\"paper\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"A_28547103.0\",\"xref\":\"x\",\"y\":0.8,\"yref\":\"y\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"G_28547122.0\",\"xref\":\"x\",\"y\":0.8,\"yref\":\"y\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"C_28545668.0\",\"xref\":\"x3\",\"y\":0.8,\"yref\":\"y3\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"T_28545687.0\",\"xref\":\"x3\",\"y\":0.8,\"yref\":\"y3\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"A_28546080.0\",\"xref\":\"x5\",\"y\":0.8,\"yref\":\"y5\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"T_28546175.0\",\"xref\":\"x5\",\"y\":0.8,\"yref\":\"y5\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"A_28545761.0\",\"xref\":\"x7\",\"y\":0.8,\"yref\":\"y7\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"G_28545850.0\",\"xref\":\"x7\",\"y\":0.8,\"yref\":\"y7\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"T_28547184.0\",\"xref\":\"x9\",\"y\":0.8,\"yref\":\"y9\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"C_28547425.0\",\"xref\":\"x9\",\"y\":0.8,\"yref\":\"y9\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"C_28547116.0\",\"xref\":\"x11\",\"y\":0.8,\"yref\":\"y11\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"C_28547135.0\",\"xref\":\"x11\",\"y\":0.8,\"yref\":\"y11\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"C_28547428.0\",\"xref\":\"x2\",\"y\":0.8,\"yref\":\"y2\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"G_28547187.0\",\"xref\":\"x2\",\"y\":0.8,\"yref\":\"y2\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"C_28545850.0\",\"xref\":\"x4\",\"y\":0.8,\"yref\":\"y4\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"T_28545761.0\",\"xref\":\"x4\",\"y\":0.8,\"yref\":\"y4\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"G_28546217.0\",\"xref\":\"x6\",\"y\":0.8,\"yref\":\"y6\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"T_28546198.0\",\"xref\":\"x6\",\"y\":0.8,\"yref\":\"y6\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"T_28545905.0\",\"xref\":\"x8\",\"y\":0.8,\"yref\":\"y8\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"G_28545886.0\",\"xref\":\"x8\",\"y\":0.8,\"yref\":\"y8\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"A_28547471.0\",\"xref\":\"x10\",\"y\":0.8,\"yref\":\"y10\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"G_28547452.0\",\"xref\":\"x10\",\"y\":0.8,\"yref\":\"y10\"},{\"showarrow\":false,\"text\":\"5'\",\"x\":\"G_28547425.0\",\"xref\":\"x12\",\"y\":0.8,\"yref\":\"y12\"},{\"showarrow\":false,\"text\":\"3'\",\"x\":\"A_28547184.0\",\"xref\":\"x12\",\"y\":0.8,\"yref\":\"y12\"}],\"title\":{\"text\":\"vgsc primer pairs | AG1000G-GH | target AGAP006227-RA\",\"x\":0.5},\"height\":1600,\"width\":1600,\"showlegend\":false},                        {\"responsive\": true}                    ).then(function(){\n",
              "                            \n",
              "var gd = document.getElementById('42917f08-771c-4bdb-8260-f0ce5202d774');\n",
              "var x = new MutationObserver(function (mutations, observer) {{\n",
              "        var display = window.getComputedStyle(gd).display;\n",
              "        if (!display || display === 'none') {{\n",
              "            console.log([gd, 'removed!']);\n",
              "            Plotly.purge(gd);\n",
              "            observer.disconnect();\n",
              "        }}\n",
              "}});\n",
              "\n",
              "// Listen for the removal of the full notebook cells\n",
              "var notebookContainer = gd.closest('#notebook-container');\n",
              "if (notebookContainer) {{\n",
              "    x.observe(notebookContainer, {childList: true});\n",
              "}}\n",
              "\n",
              "// Listen for the clearing of the current output cell\n",
              "var outputEl = gd.closest('.output');\n",
              "if (outputEl) {{\n",
              "    x.observe(outputEl, {childList: true});\n",
              "}}\n",
              "\n",
              "                        })                };                            </script>        </div>\n",
              "</body>\n",
              "</html>"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 158,
      "id": "shXKBpG49LSU",
      "metadata": {
        "id": "shXKBpG49LSU",
        "outputId": "727d214a-371e-4d9a-812e-d80d822d5ec8",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply site filters if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply sample query if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply site filters if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply sample query if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply site filters if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply sample query if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply site filters if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply sample query if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply site filters if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply sample query if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple regions\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply site filters if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: apply sample query if requested\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise parameters\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: normalise region to list to simplify concatenation logic\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate multiple sample sets and/or contigs\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: concatenate data from multiple sample sets\n",
            "DEBUG:malariagen_data.ag3:snp_genotypes: locate region - do this only once, optimisation\n",
            "DEBUG:malariagen_data.ag3:snp_sites: access SNP sites and concatenate over regions\n",
            "DEBUG:malariagen_data.ag3:snp_sites: apply site mask if requested\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-158-b4d702010a31>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m                                                 \u001b[0massay_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0massay_type\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m                                                 \u001b[0mseq_parameters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mseq_parameters\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m                                                 save=True)\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/AgamPrimer/AgamPrimer.py\u001b[0m in \u001b[0;36mplot_primer_ag3_frequencies\u001b[0;34m(primer_df, gdna_pos, contig, sample_set, assay_type, seq_parameters, save, sample_query)\u001b[0m\n\u001b[1;32m    288\u001b[0m   \u001b[0;31m# Loop through each primer pair and get the frequencies of alternate alleles, storing in dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m   \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_primer_pairs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 290\u001b[0;31m     \u001b[0mres_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_primer_alt_frequencies\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprimer_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgdna_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_set\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massay_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_query\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m   \u001b[0;31m# Plot data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/AgamPrimer/AgamPrimer.py\u001b[0m in \u001b[0;36mget_primer_alt_frequencies\u001b[0;34m(primer_df, gdna_pos, pair, sample_set, assay_type, contig, sample_query)\u001b[0m\n\u001b[1;32m    195\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    196\u001b[0m   \u001b[0moligos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreturn_oligo_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0massay_type\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 197\u001b[0;31m   \u001b[0mfreq_arr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref_arr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpos_arr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_primer_arrays\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcontig\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcontig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgdna_pos\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgdna_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_set\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msample_set\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massay_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0massay_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_query\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msample_query\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    198\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    199\u001b[0m   \u001b[0mdi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/AgamPrimer/AgamPrimer.py\u001b[0m in \u001b[0;36mget_primer_arrays\u001b[0;34m(contig, gdna_pos, sample_set, assay_type, sample_query)\u001b[0m\n\u001b[1;32m    179\u001b[0m       \u001b[0mspan_str\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf'{contig}:{span[0]}-{span[1]}'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    180\u001b[0m       \u001b[0mprimer_ref_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mag3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenome_sequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mspan_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m     \u001b[0;31m# get dna sequence for span\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 181\u001b[0;31m       \u001b[0mgeno\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mag3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msnp_genotypes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mspan_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_sets\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msample_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m     \u001b[0;31m# get genotypes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    182\u001b[0m       \u001b[0mfreqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mallel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenotypeArray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgeno\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount_alleles\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_frequencies\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m                                \u001b[0;31m# calculate allele frequencies\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    183\u001b[0m       \u001b[0mfreq_arr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfreq_arr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfreqs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/malariagen_data/ag3.py\u001b[0m in \u001b[0;36msnp_genotypes\u001b[0;34m(self, region, sample_sets, sample_query, field, site_mask, inline_array, chunks)\u001b[0m\n\u001b[1;32m   1107\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1108\u001b[0m                 \u001b[0mpos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msnp_sites\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfield\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"POS\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1109\u001b[0;31m                 \u001b[0mloc_region\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlocate_region\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1110\u001b[0m                 \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mloc_region\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1111\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/malariagen_data/util.py\u001b[0m in \u001b[0;36mlocate_region\u001b[0;34m(region, pos)\u001b[0m\n\u001b[1;32m    438\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    439\u001b[0m     \"\"\"\n\u001b[0;32m--> 440\u001b[0;31m     \u001b[0mpos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mallel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSortedIndex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    441\u001b[0m     \u001b[0mloc_region\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocate_range\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mregion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mregion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    442\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mloc_region\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/allel/model/ndarray.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, copy, **kwargs)\u001b[0m\n\u001b[1;32m   3377\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3378\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3379\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSortedIndex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3380\u001b[0m         \u001b[0mcheck_ndim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3381\u001b[0m         \u001b[0;31m# check sorted ascending\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/allel/model/ndarray.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, copy, **kwargs)\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     75\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 76\u001b[0;31m         \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     77\u001b[0m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNumpyArrayWrapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/dask/array/core.py\u001b[0m in \u001b[0;36m__array__\u001b[0;34m(self, dtype, **kwargs)\u001b[0m\n\u001b[1;32m   1624\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1625\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__array__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1626\u001b[0;31m         \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1627\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdtype\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1628\u001b[0m             \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0mdask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m         \"\"\"\n\u001b[0;32m--> 290\u001b[0;31m         \u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    291\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mschedule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdsk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 574\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrepack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpostcomputes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    575\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/dask/base.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mschedule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdsk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 574\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrepack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpostcomputes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    575\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/dask/array/core.py\u001b[0m in \u001b[0;36mfinalize\u001b[0;34m(results)\u001b[0m\n\u001b[1;32m   1207\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mtuple\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1208\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1209\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mconcatenate3\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1210\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1211\u001b[0m             \u001b[0mresults2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresults2\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/dask/array/core.py\u001b[0m in \u001b[0;36mconcatenate3\u001b[0;34m(arrays)\u001b[0m\n\u001b[1;32m   5084\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mndim\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5085\u001b[0m                 \u001b[0marr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m...\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5086\u001b[0;31m         \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5087\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5088\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ],
      "source": [
        "results_dict = AgamPrimer.plot_primer_ag3_frequencies(primer_df=primer_df,\n",
        "                                                gdna_pos=gdna_pos,\n",
        "                                                contig=contig,\n",
        "                                                sample_set=sample_set, \n",
        "                                                sample_query=sample_query,\n",
        "                                                assay_type=assay_type,\n",
        "                                                seq_parameters=seq_parameters,\n",
        "                                                save=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "aCVuE1CiAR4h",
      "metadata": {
        "id": "aCVuE1CiAR4h"
      },
      "source": [
        "Now lets plot these primer pairs across the genome, highlighting where they are in relation to any nearby exons..."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "cAacbw-q8Sgm",
      "metadata": {
        "id": "cAacbw-q8Sgm"
      },
      "outputs": [],
      "source": [
        "AgamPrimer.plot_primer_locs(primer_res_dict=results_dict, assay_type=assay_type, gff=ag3.geneset(), contig=contig, seq_parameters=seq_parameters, n_primer_pairs=n_primer_pairs, legend_loc='lower left')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0JA92ApBcxLo",
      "metadata": {
        "id": "0JA92ApBcxLo"
      },
      "outputs": [],
      "source": [
        "primer_df"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "g4r9sQectXxx",
      "metadata": {
        "id": "g4r9sQectXxx"
      },
      "source": [
        "####**We may now have designed suitable primers. However, there are some further considerations...**\n",
        "\n",
        "\n",
        "- Primers **must** be run in [**NCBI Blast**](https://blast.ncbi.nlm.nih.gov/Blast.cgi), to ensure specificity against the host organism, and specificity for the genomic location of interest. I would recommend both doing a general blast and also specifically against *An. gambiae* (TaxonID = 7165).\n",
        "\n",
        "- If in multiplexed use with other primers or probes, primers must not interact with each other. This can be investigated on a one by one basis using the IDT tool [oligoanalyzer](https://eu.idtdna.com/calc/analyzer), though higher throughput algorithms may be required.\n",
        "\n",
        "- If designing Locked Nucleic Acid (LNA) probes for SNP detection, you will want to play around with the placement of LNAs in the oligo sequence, which can allow short probes (~10-14 bases) to bind with high affinity and discriminate between SNPs. IDT have a tool for this which allow you to check the binding affinity between mismatches, though it requires a log in https://eu.idtdna.com/calc/analyzer/lna. \n",
        "\n",
        "- Many more considerations.... [IDT - How to design primers and probes for PCR and qPCR](https://eu.idtdna.com/pages/education/decoded/article/designing-pcr-primers-and-probes)   \n",
        "\n",
        "\\\n",
        "  \n",
        "\n",
        "####**Future development**\n",
        "\n",
        "Any contributions or suggestions on how we can improve this notebook are more than welcome. Please [email](mailto:sanjay.nagi@lstmed.ac.uk) or log an [issue on github](https://github.com/sanjaynagi/primerDesignAg/issues). This notebook and source code for AgamPrimer are located here - https://github.com/sanjaynagi/AgamPrimer/    \n",
        "\n",
        "\\\n",
        "####**References**\n",
        "\n",
        "The Anopheles gambiae 1000 Genomes Consortium (2020). **Genome variation and population structure among 1142 mosquitoes of the African malaria vector species *Anopheles gambiae* and *Anopheles coluzzii***. *Genome Research*, **30**: 1533-1546. \n",
        "https://genome.cshlp.org/content/early/2020/09/25/gr.262790.120\n",
        "\n",
        "Untergasser A, Cutcutache I, Koressaar T, Ye J, Faircloth BC, Remm M and Rozen SG (2012). **Primer3--new capabilities and interfaces**. *Nucleic Acids Research*. 40(15):e115."
      ]
    }
  ],
  "metadata": {
    "colab": {
      "name": "Primer Design in Anopheles gambiae.ipynb",
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.12"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "1c032b889e4348b997161fd4a9f4e113": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_e6904bebdccc4e4689867a6826a46fd2",
              "IPY_MODEL_ef41c110036f497fa80560c554b01cd5",
              "IPY_MODEL_ede22965068044cca74f1b09c31b8ed9"
            ],
            "layout": "IPY_MODEL_60e494897ca448cbb59e98dd440f76b7"
          }
        },
        "e6904bebdccc4e4689867a6826a46fd2": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_3ce652ccf9984bb7817df8b110097b34",
            "placeholder": "​",
            "style": "IPY_MODEL_2d36054a85bb4d92b97d38f71f38207d",
            "value": "Load sample metadata: 100%"
          }
        },
        "ef41c110036f497fa80560c554b01cd5": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c4e3923c21a9460d83ff7914623db3a6",
            "max": 28,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_14df1c19ee744bf38b7e131c3f5e79ae",
            "value": 28
          }
        },
        "ede22965068044cca74f1b09c31b8ed9": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_2d294c504fdc4d9480fe90b3465b4fdf",
            "placeholder": "​",
            "style": "IPY_MODEL_ecc9e37d989b4651962ff368a1659340",
            "value": " 28/28 [00:05&lt;00:00,  5.12it/s]"
          }
        },
        "60e494897ca448cbb59e98dd440f76b7": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "3ce652ccf9984bb7817df8b110097b34": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "2d36054a85bb4d92b97d38f71f38207d": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "c4e3923c21a9460d83ff7914623db3a6": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "14df1c19ee744bf38b7e131c3f5e79ae": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "2d294c504fdc4d9480fe90b3465b4fdf": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ecc9e37d989b4651962ff368a1659340": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}